KEP-3488: CEL Admission Control (#3492)
kubernetes/enhancements: keps/sig-api-machinery/3488-cel-admission-control/README.md

## Background

This is not a new idea, Tristan Swadell (@TristonianJones) explored policy for
Kubernetes using CEL with
[cel-policy-templates-go](https://github.com/google/cel-policy-templates-go),
and Jordan Liggitt (@liggitt) prototyped using CEL for in-process admission
control in Kubernetes in 2020.

[CRD Validation
Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
were implemented as a more constrained subset of this problem and addressed how
to integrate the Kubernetes type system with CEL.

## Considerations

### Admission Webhook Parity

Users of the Kubernetes API are already familiar with
ValidatingWebhookConfigurations. We should strive for consistency with this API
unless there is a good reason to diverge from it.  As a concrete example, we
should provide access to all the information that webhooks have access to (see
[AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
and if we provide access to additional information we should extend

> AdmissionRequest to include it.

Author: I've added some caveats to this goal. It sounds like we could safely add a few fields (with negligible impact to object size) but that we should avoid (or be very careful) adding anything large, and that things the webhook could retrieve with an informer (like namespace objects) might not ever need to go into AdmissionReview.

to have policy frameworks generating a CEL expression for each possible
configuration.

### Migration

With webhooks already in large scale use in the Kubernetes ecosystem, we intend
to prioritize capabilities that ease migration. As a concrete example, when
migrating, having fine grained control of what validation messages are returned
and how they are formatted can make a migration far more seamless.

### Compliance

In-process admission control has fundamental advantages over webhooks: it is far
safer to use in a "fail closed" mode. With webhooks, using "fail closed" can
negatively impact cluster availability. But "fail closed" is very valuable when
enforcing compliance (and security). We intend to prioritize capabilities that
make "fail closed" a safe mode of operation. As a concrete example, only
allowing CEL expressions that pass compilation and type checking significantly
reduces the opportunities for runtime errors.

Also, making it possible (and convenient) to declare zero trust policies is
important to compliance. i.e., it should be possible for policies governing 
resources like namespace and roles to default to the most restrictive state

> when first created.

Author: This is a good list. I've threaded them into this KEP.  "static or on-initialization" I listed as a non-goal since is seems that it should be a generically added feature not specific to admission. The other's I've grouped in with some of the safety feature TODOs in phase 1 and 2. I'm not sure which ones we will commit to yet, but we shouldn't loose track of them.

## Proposal

### Phase 1

Introduce a new "ValidatingAdmissionConfiguration" kind to the
admissionregistration.k8s.io group. (suggestions welcome on exact name to use
for kind)

At a high level, the API needs to support:

- Request matching (similar to webhook Rules, NamespaceSelector,
  ObjectSelector)
- CEL rule evaluation (similar to both [CRD Validation
  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
  and
  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))

> - Version conversion support (similar to webhooks MatchPolicy)

Author: Ah, that's a good point. I agree with this.

- Access the old object (similar to [transition
  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
  and oldObject in AdmissionRequest)
- Configurability, as motivated above.

There are also lots of details (response message formatting, severity
levels, failure policies, type safety, ...) that will be discussed in detail
further on in this proposal.

#### Configurability in the API

How configurability is expressed in the API will have significant influence
how the API is shaped. We are currently considering two design options:

##### Configurability Option 1: Embed configuration into the object, grouped with match criteria

```
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionConfiguration
metadata:
  name: "validate-xyz.example.com"
configurationSchema:  

>   openAPIV3Schema: ... # the below configuration is validated against this schema

Author: We've changed what design we'd like to propose for this to instead reference a CRD like done by the [Dynamic Resource Allocation KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation). I've also written up some of the alternatives considered.

spec:
  match:

>   - rules: ... # see ValidatatingWebhookConfiguration rules

Author: See the "Type safety" section. I've added some details about alternatives considered.  This is one of the more critical decisions we will need to make in this KEP, so appreciate all feedback on it.

    objectSelectors: ...  # "
    namespaceSelectors ... # "
    configuration: {x: 1} # each match may be configured differently
  - rules: ...
    objectSelectors: ...
    namespaceSelectors ...
    configuration: {x: 2}
validations:
  - rule: "self.spec.xyz == configuration.x"
    # ...other rule related fields here...
```

Pros:

- Self-contained resource
- Higher level policy abstractions can be defined out-of-tree and can be
  translated into this form by controllers

Cons:

- The resource grows by the number of match rules and amounts of configuration data.

> - Need to define how configurations of overlapping matches are interpreted. If

Reviewer: not exactly related to the con noted here, but this seemed a useful place to mention prior art references for structure of an ordered set of matching rules:
* rbac, priority & fairness, and admission webhook configurations all use ordered lists of rules, and first match allows/matches
* audit configuration also uses an ordered list of rules, but can make a match exclude auditing... this approach could address the "ability to skip specific resource types" mentioned above

Author: Thx! Added. We're still nailing down what exact match rules we want to make available.

  any of the matches fail does the rule as an admission check fail, or if any
  matches pass does the admission check as a whole pass?

##### Configurability Option 2: Put configuration in separate resources

```
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionConfiguration
metadata:
  name: "validate-xyz.example.com"
spec:

>   configurationCustomResourceDefinition:

Reviewer: I can't tell, is this inlining a CRD, or is an object reference?

Author: Inline CRD.  But we've since updated the KEP to propose using an object reference very similar to what https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#theory-of-operation does.  I've fleshed out some of the KEP explaining the rationale and alternatives.  This part of the KEP is still under active discussion.

    group: rules.example.com
    kind: LimitRule
    ...
      openAPIV3Schema: ... # the below configuration is validated against this schema
  validations:
    - rule: "self.spec.xyz == configuration.x"
      # ...other rule related fields here...
    ...
---
apiVersion: rules.example.com/v1
kind: LimitRule
metadata:

>   name: "validate-xyz-rule1.example.com"

Reviewer: is the name required to match the ValidatingAdmissionConfiguration and be a singleton?

Author: No, multiple configuration objects are allowed. Updated KEP tries to explain this better.

spec:
  match:
  - rules: ...
    objectSelectors: …

>     namespaceSelectors ... 

Reviewer: was the match/rules portion supposed to be included in this custom resource, or just the `x: 1` bit?

if so, separating the validation rules from the match rules seems like it would make it *really* hard to type check and understand the safety of changes to either object

Author: Matching and configuration both go into the custom resource.

This allows policy to be configured differently for different resources.  I've added more examples of this in the updated KEP.

re: type safety.  Yes. My current thinking is that the policy definition (`ValidatingAdmissionPolicy`) needs to establish a constraint on which GVKs it can validated so that type-checking can be performed and then all the custom resources that configure the policy need to be constrained to that GVK. I'm working on clearing this up in the KEP.

  - CRDs need to be created when a ValidatingAdmissionConfiguration is created
  - Requires the API server establish watches on the CRs to accumulate all the
    configuration information
- Possible to implement using Option 1. A controller can construct a
  resource for the API in Option 1 from custom resources that define this
  option (e.g. via a controller).
- CRD lifecycle prevents CRs from being written until the CRD is written, which
  means that until a ValidatingAdmissionConfiguration is written AND a CRD is
  created for it, no CRs can be written.  Applying multiple resources
  (e.g. kubectl apply -f <lots of yaml files>) only works if the apply
  operations are ordered and objects are verified to exist at specific steps.

#### Additional Capabilities (WIP)

In addition to the high level API structure proposed above, there are quite
a few capatiblities that are important to this enhancement that we are working
on defining. These are all work-in-progress.

Rule Message:

- message formatting (need support for string formatting. Allow a CEL expression
  to be used to format the whole string? Or allow CEL expressions to evaluate
  template values?)

> - severity (e.g. "Warning")

Reviewer 2: Assuming rules can be configured to populate https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings via AdmissionResponse rather than reject a request, is there also a use case for another class of warning that isn't exposed to clients?

Reviewer: > is there also a use case for another class of warning that isn't exposed to clients?

That sounds like audit annotations, which can be returned by admission webhooks today.

I think the info included in validating admission webhook responses covers this:
* warnings
* audit annotations
* admission decision, one of:
  * accept
  * reject (optionally with a custom status code, reason, message, details, etc)

Reviewer: (note specifically that warnings and audit annotations can be returned for allowed requests)

- status type equivalent (e.g. Forbidden can be returned from a webhook using
  HTTP 403) (can a match rule override this? or can it be selected via a CEL
  expression?)
- Similar to how each match has configuration, can each match also select
  the severity of messages?

Rule scope:

> - Dereferencing to deeply nested fields in CEL can be burdonsome. A field

Reviewer: ```suggestion
- Dereferencing to deeply nested fields in CEL can be burdensome. A field
```

>   containing a field path would allow a CEL rule to be scope to some node in

Reviewer: ```suggestion
  containing a field path would allow a CEL rule to be scoped to some node in
```

  the schema.
- Also, the idea of named scopes (e.g. "ALL_CONTAINERS") would allow a single
  rule to be applied to all container definitions in a resource (e.g. both

>   "containers" and "initContainers" in a pod)

Reviewer: and ephemeralContainers (seemingly inconsequential nit which actually illustrates the importance of an "all containers" helper since everyone forgets them :)

Author: Happy to make this list longer to emphasize the point :)

Failure Policy:

- Not the same as webhooks, where failure policy is about the remote request
- What if a CEL rule evaluates to an error?
- What if a CEL rule is for a CRD kind and the CRD changes such that the CEL
  rule fails type checking? (or the CRD does not exist?)

Decision Rules:

- Are overlapping matches combined using an OR or AND? Can this be configured?
- Can a validation rule depend on another validation rule (only evaluate if the
  other rule is true?)
- How to make it so that an "not authorized" error prevents all other validation
  messages from being returned (to prevent exfiltration)?

Composition utilities:

- Ability to define a sub-expression and then use it in validation rules (xref
  cel-policy-template "terms")

Transition rules:

> - Should we use "self" and "oldSelf" ? How do transition rules work for delete?

Reviewer: as much as possible, I would encourage matching admission webhook AdmissionReview semantics. `object` is null for DELETE operations, and `oldObject` is null for `CREATE` operations

Author: +1, I'm in favor of using "object" as the variable throughout this KEP (instead of "self"). That keeps things consistent with AdmissionReview.  I've incorporated this.

> Access to namespace labels:

Reviewer: (and probably namespace annotations)

- Most heavily needed fields not directly available in the resource being
  validated. Note that namespaceSelectors already allow matches to examine
  namespace levels.

Type safety:

- What happens when a ValidatingAdmissionConfiguration is created in a cluster
  where either CRDs or native types are incompatible with CEL rules and the CEL
  rules fail to compile?
  - Option: Put something into status about the type check error and apply the

>     fail policy instead of evaluating the CEL expression at runtime

Reviewer: status on API server configuration objects has been tricky to design in the past, because of the following:
* multiple active kube-apiservers (sometimes at identical versions, sometimes skewed by one version during upgrade)
* multiple active non-kube-apiserver servers (aggregated servers)

Author: Looks 'NonStructural' relies on the fact that the status change is a function of the resource then generation can be used (https://github.com/kubernetes/kubernetes/commit/2cfc3c69dc7c17b2711af0168f39ed7f515675c2). That's not the case here, we want to have both the CEL rules (defined in one resource) and the CRD schema defined in another resource (or native type schema, which is version dependent).

I'll make note of this in the KEP.

  - Option: Put something into status and run the CEL expression dynamically, if
    it fails with an error apply the fail policy
  - Option: Provide a field on each rule that determines the behavior
    ("TypeCheckFailPolicy: (FailAll|RunDynamic)").

> 

Reviewer: a section on interaction with aggregated API servers would be useful

1. the API server validating/persisting the ValidatingAdmissionConfiguration instances isn't the same one serving aggregated types, so wouldn't necessarily have schema info to check type safety
2. the aggregated API server is responsible for enforcing admission on its custom types, so the implementation that reads ValidatingAdmissionConfiguration instances and enforces them would have to live in k8s.io/apiserver and be active in aggregated API servers to enforce admission on aggregated types effectively (same as admission webhooks today)

Author: Thanks! I've added a section. It's still marked as TODO but it will make sure we track this.  Throws a wrench into type safety for sure, I need to think about that one.

can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  config:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  kind:
    group: apps
    kind: Deployment
    version: v1
  validations:
    - expression: "self.spec.replicas <= config.maxReplicas"
      # ...other rule related fields here...
```

The `spec.config` field of the `ValidatingAdmissionPolicy` references the
"configuration CRD" for this admission policy. For this example, the
configuration CRD is `ReplicaLimit`.

Note: This is a "Bring Your Own CRD" design. The admission policy definition

> author is responsible for providing the `ReplicaLimit` configuration CRD.

Reviewer 3: +1 to figuring out how best to deal with the shared ownership here -- to the extent that the schema for the configuration CRD is configuring functionality implemented by K8s, figuring out how to deal with issues like version skew will be important.

> author is responsible for providing the `ReplicaLimit` configuration CRD.

Author: After looking more at version skew issues, CRD issues and aggregated API server issues, I'm incline to make CEL dynamic, we can still run the type checker but it will only be informational (generates warnings). This helps with part of the problem. The other thing we can do is use matchPolicy, I've added "Use Case: Multiple policy definitions for different versions of CRD" to show an example of this idea.

}
```

Example usage:

```go
// +genclient
// +genclient:nonNamespaced
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
type ReplicaLimit struct {
  MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,1,name=maxReplicas"`

  Match Match `json:"match,omitempty" protobuf:"bytes,2,name=match"`
}
```

TODO: What exact matching mechanisms will be supported? Right now this KEP
shows the use of the same match rules that admission webhooks use.

TODO: Should policy definitions also have match criteria? Match criteria at the
policy definition level could be used constrain what resources the policy may
apply to across the entire cluster (like the above "kind constraint" but more
generally). This would allow policy authors to constrain use of the policy to
the resources it was designed for.

TODO: Should "standalone policies" be supported? I.e. should it be possible to
declare a policy definition that has no `spec.config` but has it's own match
criteria? Right now a policy requires a minimum of three resources to do
anything (policy definition, configuration CRD, at least one policy

> configuration).

Reviewer 4: if a policy doesn't have configuration CRD or policy configuration, does that automatically make it a standalone? can a policy definition be a standalone (default validation) while supporting different parameters if there are policy configurations? e.g. in the example above, standalone defaults to maxReplicas of 100 unless additional policy configurations are deployed that overrides the default based on the match criteria?

> configuration).

Author: I've added a stawman proposal for standalone policy configurations (I called them "singleton policies" in the KEP).  I elected to add a field that explicitly toggles the state of a policy from "Configurable" to "Singleton" just for added clarity.

> configuration).

Author: > can a policy definition be a standalone (default validation) while supporting different parameters if there are policy configurations?

There are use cases where a default (least privilege) policy is needed. I originally had wanted to have the policy definition provide the default for cases like this. But that can be problematic. If a default configuration that is intended to be configurable on a per cluster basis is nested in the policy definition, the permissions to configure the defaults are the same as the permissions to modify the policy definition.

I've added a "Exclude and default matching" to cover this and added a bunch of options. I'm not happy with any of them but extra eyes on that would be very welcome.

in detail further on in this proposal.

We have divided this proposal into phases, all of which must be completed before
this feature graduates to beta. Our goal is to size the phases so that each
can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  config:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  constraints:
    rules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
      scope:       "*"
  validations:

>     - expression: "object.spec.replicas <= config.maxReplicas"

Reviewer 5: `expression` should have a `language` field.

e.g.

```yaml
validations:
- expression: "object.spec.replicas <= config.maxReplicas"
  language: cel
...
```

The reason for this is to support future additional languages that we may decide to add in the future. Implicit expression languages can be defaulted in, but will be present in the object represantion. So it doesn't hinder usability, but it does make the API object more future-proof/flexible.

Author: For CEL Validation Rules, we had a small section about additional language support: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#other-expression-languages

The idea was that when we introduce a second language, we add the field to allow the user to pick it. Since we always have a expression defined in a struct where the language field can be added later, we're future proofed.

```

The `spec.config` field of the `ValidatingAdmissionPolicy` references the
"configuration CRD" for this admission policy. For this example, the
configuration CRD is `ReplicaLimit`.

Note: This is a "Bring Your Own CRD" design. The admission policy definition
author is responsible for providing the `ReplicaLimit` configuration CRD.

`spec.kind` specifies that the CEL expressions in `validations.expression` have
access to the fields of the `v1.Deployment` kind.

To configure an admission policy, "policy configurations" of the configuration
CRD kind are created.

For example:

```yaml
# Policy configurations
apiVersion: rules.example.com/v1
kind: ReplicaLimit
metadata:
  name: "replica-limit-test.example.com"
spec:
  match:
    namespaceSelectors:
    - key: environment,
      operator: In,
      values: ["test"]
  maxReplicas: 3
```

This `replica-limit-test.example.com` policy configuration limits deployments to
a max of 3 repliacas in all namespaces in the test environment.

An admission policy may have multiple configurations. In this example an
additional `ReplicaLimit` could be created to configure the "prod" environment
to have a maxReplicas limit of 100.

> This design separates admission policy _definition_ from _configuration_. This

Reviewer 4: Is the _configuration_ cluster scoped or namespace scoped? 
Might be good to add a snippet about how we plan to do versioning for the policy _definition_.

  `ValidatingAdmissionPolicy`, configuration CRDs and configuration resources
  are written to the API Server. This provides a workflow where type-checking
  can be used to detect errors early, preventing them from occurring during
  admission control validation and making it safer to operate a policy using
  `failureMode: Fail`.

#### Failure Policy

For in-process validation there is no remote request, so errors should be
deterministic. We believe this significantly reduces the risk of "fail closed"
admission control as compared to webhooks.

Because failure policy is most often selected based on the need to guarantee
enforcement, we propose defaulting failure policy to "fail" and allowing it to
be configured on a per-rule basis:

```yaml
  validations:
    - rule: "object.spec.xyz == configuration.x"
      failurePolicy: Ignore # The default is "Fail"
```

TODO: Metric for ignored failures?

#### Safety measures

To prevent clusters from being put into a unustable state that cannot be
recoverd from via the API, admission webhooks are not allowed to match
`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.

> We will extend this approach, and not allow `ValidatingAdmissionPolicy`,

Reviewer 5: I think that this is problematic. It makes sense for `ValidatingAdmissionPolicy` to not be applied to itself, but there's no reason why they can't apply to `ValidatingWebhookConfiguration` (or vice versa)

e.g. `ValidatingAdmissionPolicy` applies to everything except `ValidatingAdmissionPolicy`, `ValidatingWebhookConfiguration` applies to everything except `ValidatingWebhookConfiguration` _and_ `ValidatingAdmissionPolicy`

This closes a gap today where `ValidatingWebhookConfigurations` can not be restricted to targetting specific hosts.

Author: I agree. I've updated the KEP to match. I suppose this approach could be generalized in the future if needed. Maybe each `ValidatingAdmissionPolicy` can specify a level and a `ValidatingAdmissionPolicy` of a lower level could match a `ValidatingAdmissionPolicy` of a higher level, eliminating the possibility of a cycle like done here between webhooks and cel.

`ValidatingWebhookConfiguration` or `MutatingWebhookConfiguration` kinds to be
matched by any of the admission control extension mechanisms (webhooks or cel).

We are also considering not allowing a `ValidatingAdmissionPolicy` to match the
configuration CRD kind that is uses for config. It would be possible to recover

> form this since the `ValidatingAdmissionPolicy` would remain accessible, but

Reviewer 5: typo: "from this"

the control granularity that PSP provided.

There are numerous in-tree embedded controllers.

The existence of security regimes like the CIS Kubernetes Benchmarks highlight
the values of standardized controls. Automating their enforcement, where
possible, will make it easier for users to lock down their clusters.

With the advent of CRDs, and the drive to make the resources they define
first-class entities, the footprint of Kubernetes extensions is set to grow for
the foreseeable future. This KEP allows authors of such extensions to provide
policy primitives similar to PSP or PSA, putting them on equal footing with
in-tree functionality.

With the reduced infrastructure footprint and demonstrated demand for a
customizable, built-in mechanism for extensible policy, this KEP fills a
community need. It is not intended to replace validating admission webhooks
altogether, however, since these can support functionality that may not make
sense to provide in-tree.

### Goals

- Provide an alternative to webhooks for the vast majority of validating
  admission use cases.
- Provide the in-tree extensions needed to build policy frameworks for

>   Kubernetes, again without requiring webhooks for the vast majority of use

Reviewer 4: Might be good to clarify what type of use cases are ideal for this in-tree extension vs a webhook

unless there is a good reason to diverge from it.  As a concrete example, we
should provide access to all the information that webhooks have access to (see
[AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
and if we provide access to additional information we should extend
AdmissionRequest to include it. (But we need to be careful not to make
AdmissionRequest significantly larger as this will impact the
performance/latency of existing webhooks not leveraging the additional
information.  We should be careful about providing access to cross-object
information like namespace objects to webhooks since they can be stale.)

### Configurability

Consider an admission rule that disallows requests based on a blocklist.

While it is possible to inline a blocklist directly into a CEL expression as a
data literal (`!(object.metadata.name in ['blocked1', 'blocked2'])`), this quickly
becomes problematic:

- Long blocklists become unwieldy quickly in CEL expressions
- A blocklist per scope (e.g. namespace) is inconvenient express and maintain

> The need to configure admission rules is common enough (see below use cases)

Reviewer 6: Another common way to support this is to perform secondary authorization checks to the kube-apiserver authorizer to allow the configuration to happen in an RBAC rule. The `use` verb for PSP or openshift SCC is an example of this.

make "fail closed" a safe mode of operation. As a concrete example, only
allowing CEL expressions that pass compilation and type checking significantly
reduces the opportunities for runtime errors.

Also, making it possible (and convenient) to declare "zero trust" policies is
important to compliance. By "zero trust", we mean policy rules that apply the
principle of least privilege to newly created resources (e.g. namespaces) where
the policy is initially set to the most restrictive state and can be made less
restrictive via configuration.

## Proposal

Introduce a new `ValidatingAdmissionPolicy` kind to the
admissionregistration.k8s.io group. (suggestions welcome on exact name to use
for kind)

At a high level, the API will support:

- Request matching (similar to webhook Rules, NamespaceSelector,
  ObjectSelector)
- CEL rule evaluation (similar to both [CRD Validation
  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
  and
  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))

> - Automatic version conversion support (similar to webhooks MatchPolicy=Equivalent), with ability to introspect GVK of admitted resource in CEL expression.

Reviewer 6: I'm in favor of providing this, but will there be some kind of feedback in the resource that indicates this is happening?  Having conversion means that webhook calls enter into the flow.

Reviewer 3: Another thing to think about is portability -- conversion webhooks aren't always available shift left/right.

It might be worth supporting both conversion and the ability to write policies against multiple versions of the same object.

Author: Added to KEP.

We have divided this proposal into phases, all of which must be completed before
this feature graduates to beta. Our goal is to size the phases so that each
can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  config:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  constraints:
    rules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
      scope:       "*"
  validations:
    - expression: "object.spec.replicas <= config.maxReplicas"
      # ...other rule related fields here...
```

> The `spec.config` field of the `ValidatingAdmissionPolicy` references the

Reviewer 6: Above in the doc I see

> only allowing CEL expressions that pass compilation and type checking significantly reduces the opportunities for runtime errors.

If the schema for the config is located elsewhere (CRD I suspect), how will this be done?  I might be ok not doing it, but I'd like to understand the limitations of what we're providing.

Author: The basic idea (which gets complicated when we dig into details) is to do something to CRD registration where each API server watches all validatingAdmissionPolicy resources and for each policy resource update, reads the configuration CRD and type checks against it.

There can be races. CR validation  has a very similar race. Special case: CRD doesn't exist yet. For that one we'd need to fail type checking until the CRD becomes available.

Reporting type check errors is problematic. We can put them in the status, but since this is an API server configuration object, we don't have a leader elected controller so it's tricky to write to status.

Reviewer 4: This was a challenge for Gatekeeper as well but we relied on the status field of constraints and constraint templates. 

> We can put them in the status, but since this is an API server configuration object, we don't have a leader elected controller so it's tricky to write to status.

Does this mean this is still TBD?

Author: Yes, I'd characterize this part of the design as incomplete or TBD. I _really_ want to offer typesafety in this design, to whatever extent is possible. But it hinges on working out design details like this.

Author: We've added a "Writing to Status" section that I hope resolves this.

"configuration CRD" for this admission policy. For this example, the
configuration CRD is `ReplicaLimit`.

Note: This is a "Bring Your Own CRD" design. The admission policy definition

> author is responsible for providing the `ReplicaLimit` configuration CRD.

Reviewer 6: Will there be any automatic admission on CRDs to ensure the schemas when modified conform to the expectations of the ValidatingAdmissionPolicy?

Author: I wasn't planning on using admission control.  I want to support CRDs and ValidatingAdmissionPolicies being created in any order.  If the ValidatingAdmissionPolicy is created first, it will need to exist in a "misconfigured" state until the CRD is created.

`spec.constraints` specifies that the CEL expressions in
`validations.expression` have access to the fields of the `v1.Deployment` kind.
Multiple resources of the same kind may be matched, e.g. `*/scale`.

To configure an admission policy, "policy configurations" of the configuration
CRD kind are created. For example:

> ```yaml

Reviewer 6: What happens if I have two different ReplicaLimit objects that apply to a given namespace.  Say one selects on environment and the other selects on coolness-of-this-user?

Reviewer 3: I'd imagine every ReplicaLimit object that matches the current object-under-test would be evaluated -- they'd all have to be satisfied for the request to be allowed.

Author: Correct, configurations can have overlapping match criteria. The behavior will be that the policy is evaluated for each matching configuration. For example, if you have a blocklist policy and two configurations (each containing their own set of blocklist entries) overlap, the policy is evaluated twice, once with each blocklist.  (We won't attempt to merge the blocklists, we'll just run the policy as-is with each blocklist).

Reviewer 6: > Correct, configurations can have overlapping match criteria. The behavior will be that the policy is evaluated for each matching configuration. For example, if you have a blocklist policy and two configurations (each containing their own set of blocklist entries) overlap, the policy is evaluated twice, once with each blocklist. (We won't attempt to merge the blocklists, we'll just run the policy as-is with each blocklist).

Can you indicate that in this section?

Author: Done!

# Policy configurations
apiVersion: rules.example.com/v1
kind: ReplicaLimit
metadata:
  name: "replica-limit-test.example.com"
spec:
  match:

>     namespaceSelectors:

Reviewer 6: This is inside of CRD correct?  Are we declaring a specific schema as required for configuration objects?  What if the policy objects are namespace scoped?

Reviewer 3: I don't think namespace-scoped policy objects are specifically contemplated in this KEP, but I'd imagine there'd be an implicit requirement that only objects owned by the namespace that contains the namespace-scoped policy would be subject to said namespace-scoped policy.

Author: This is discussed in the "Namespace scoped policy configuration" section (currently listed in phase 2).  The rough idea is that for any configuration CRD that is namespace scoped, each CR implicitly applies to the resources in the same namespace (and is not allows to match resources in other namespaces).

    - key: environment,
      operator: In,
      values: ["test"]
  maxReplicas: 3
```

This policy configuration limits deployments to a max of 3 repliacas in all
namespaces in the test environment.

An admission policy may have multiple configurations. To configure the "prod"
environment to have a maxReplicas limit of 100, create:

```yaml
# Policy configurations
apiVersion: rules.example.com/v1
kind: ReplicaLimit
metadata:
  name: "replica-limit-prod.example.com"
spec:
  match:
    namespaceSelectors:
    - key: environment,
      operator: In,
      values: ["prod"]
  maxReplicas: 100
```

This design separates admission policy _definition_ from _configuration_. This
has a couple advantages:

- Access to, and delegation of, policy configuration is more manageable. In
  particular, Kubernetes RBAC works well with this design.
- Without the separation, the next most obvious API shape would be to encode
  everything into a single resource, which could easily become very large and
  run into resource size limits.

##### Policy Constraints

Each `ValidatingAdmissionPolicy` resource may optionally set `spec.constraints`
to constrain the resources it validates. 

> - `spec.constraints` limits which resources may be matched configurations of the

Reviewer 4: So does this mean the `spec.match` field in the _configuration_ object will further constraint the resources in scope for this policy that the `spec.constraints` field in the `ValidatingAdmissionPolicy` object constraints?

Author: Yes. That's a good way to think about it.

  setting constraints on what can be matched, not defining match criteria
  directly.
- Use a GVK instead of match rules. This works well for type checking, but doesn't
  help establish a the criteria for what resources a policy is allowed to apply to.

##### Configuration

The `spec.config` field of the `ValidatingAdmissionPolicy` references the
"configuration CRDs" used to configure the admission policy.

Policy configuration CRDs are the interface between `ValidatingAdmissionPolicy`
authors and the administrators that configure the policies for clusters. The
configuration CRD allows policy definition authors to define how exactly a
`ValidatingAdmissionPolicy` may be configured using a OpenAPIv3 structural
schema.

Policy configuration CRDs have a few restrictions:

- Match criteria fields must be defined and must confirm to the expected schema.
  See "Match Criteria" for more details.
- For phase 1, configuration CRDs must be cluster scoped. See phase 2 (below) for
  our plan on namespace scoped configuration CRDs.

If any of the above configuration CRD restrictions are violated, the errors will
be reported in the status of the `ValidatingAdmissionPolicy`. If the match

> criteria is malformed, this unfortunately may cause the policy to fail open--

Reviewer 6: is fail open versus fail closed an option in the ValidatingAdmissionPolicy?

Author: Yes, I was thinking of putting the `failureMode` flag (defaulting to `Fail`) on each validation expression, that way they could be independently controlled without having to split out the expressions into two policy resources.

It occurred to me that there might be reasons to have `failureMode` as a field in the configuration resources, but I haven't thought it through carefully.

  skewed by one version during upgrade)
- multiple active non-kube-apiserver servers (aggregated servers)
- Note: 'NonStructural' relies on the fact that the status change is a function
  of the resource then generation can be used
  (https://github.com/kubernetes/kubernetes/commit/2cfc3c69dc7c17b2711af0168f39ed7f515675c2).
  Can this approach be somehow extended?S

Note that a policy configuration CRD may be referenced by the `spec.config` of
multiple `ValidatingAdmissionPolicy` resources. Each of which may apply
different policy validation rules using the same configuration.  For example,
one `ValidatingAdmissionPolicy` might validate the containers declared in `pods`
while another might validate the containers declared in the `podTemplate` of a
`replicaSet`. As long as both are validating that the same policy configuration
is being enforced, just in different ways, it is reasonable for both to share
the same policy configuration resources.

Alternative considered: Embed the configuration CRD schema directly into
`ValidatingAdmissionPolicy`.

Pros:

- Policy author doesn't need to define the CRD. It can instead be generated for them.

Cons:

> - This implies that the policy owns the configuration. But we have use cases

Reviewer 6: I find this pretty convincing

Reviewer 3: pretty convincing as a con?

Reviewer 6: yes, as a con

  where multiple policies share the same configuration. For this model it seems
  misleading and potentially to have both policies defining the configuration.

##### Match Criteria

During admission, the Kubernetes API server will validate the resource being
admitted against all policy configuration resources that match the resource.

Match criteria must be declared in the `spec.match` field of policy
configuration resources (see `ReplicaLimit` in the above example).

In order for policy configuration resources to declare match criteria, the

> corresponding configuration CRD schema must has a `spec.match` property. This

Reviewer 4: ```suggestion
corresponding configuration CRD schema must have a `spec.match` property. This
```

property must conform to the below "matching schema template". This ensures that
the match criteria is in the format that API server expects (the API server will
be using duck typing here since there is no established way to do polymorphism
across CRDs). The schemas of these fields in the configuration CRD may omit any
optional properties; policy definition authors should only include the parts of
the "match schema template" that are useful for configuring a particular policy.

(Also, by allowing the "matching schema template" in configuration CRDs to be a
omit optional properties, this API is future proofed against the addition of
other match related properties in the future).

"matching schema template":

```go
// TODO: Add this as a struct into the Kubernetes codebase so it can easily be
// imported?

> type Match struct {

Reviewer 6: Other common matching criteria include
1. which user or group is requesting
2. does the user or group have an authorizer permission like "use" for PSP
3. an indicator on the object being admitted that indicates which configuration to use.

Can we get a table of matching mechanisms and whether we'll support them or not?

Author: Yes, added.

  Rules []admissionregisterationv1.RuleWithOperations `json:"rules,omitempty" protobuf:"bytes,1,rep,name=rules"`
  NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
  ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,3,opt,name=objectSelector"`
}
```

Example usage:

```go
// +genclient
// +genclient:nonNamespaced
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
type ReplicaLimit struct {
  MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,1,name=maxReplicas"`

  Match Match `json:"match,omitempty" protobuf:"bytes,2,name=match"`
}
```

For comparison (provided by @liggitt):

- RBAC, priority & fairness, and admission webhook configurations all use
ordered lists of rules, and first match allows/matches.
- Audit configuration also uses an ordered list of rules, but can make a match
exclude auditing... this approach could address the "ability to skip specific
resource types" mentioned above.

> TODO: What exact matching mechanisms will be supported? Right now this KEP

Reviewer 3: prior art: Gatekeeper has match criteria that have grown from user requirements over the years:

https://github.com/open-policy-agent/gatekeeper/blob/a1add93b0beb5c48eb92a6a2eb5ee7d21551a1b6/pkg/mutation/match/match.go#L22-L56

Author: Thanks, I've added to references in match section. I consider it a requirement to have exclude support, I also like having GVK matching. Has wildcard support proved valuable?

> shows the use of the same match rules that admission webhooks use.

Reviewer 6: those match rules are effective because the plugins can choose the more granular matching criteria.

> TODO: Should policy definitions also have match criteria? Match criteria at the

Reviewer 4: +1

> TODO: Should policy definitions also have match criteria? Match criteria at the

Reviewer 3: As currently written, the KEP shows policy definitions with webhook-style "match criteria" (operation, resource, etc.)

It's worth considering whether the role of "policy match criteria" is purely for type information, or if it can be used more generally (e.g. prevent any policy from touching `kube-system`). This discussion overlaps heavily with the notion of "standalone policies" below.

Author: I've added a table that shows the capabilities of a wider range of match rules found in kubernetes (webhooks, rbac, audit, priority&fairness).  We do want to go beyond just type information.  It's clear to me that webhook match criteria doesn't do enough. Lack of exclude option, in particular, is problematic (it has made excluding kube-system difficult). Lack of resource name matching is also inconvenient.

policy definition level could be used constrain what resources the policy may
apply to across the entire cluster (like the above "kind constraint" but more
generally). This would allow policy authors to constrain use of the policy to

> the resources it was designed for.

Reviewer 4: +1

> TODO: Should "standalone policies" be supported? I.e. should it be possible to

Reviewer 6: I'm in favor of standalone policies that do not rely on configuration.  It could do things like, "this field must always be X" or "this field for this type must always be set to some value".  It could also be built to produce the effect of an RBAC deny rule for mutation.

declare a policy definition that has no `spec.config` but has it's own match
criteria? Right now a policy requires a minimum of three resources to do
anything (policy definition, configuration CRD, at least one policy
configuration).

`MatchPolicy` will implicitly be `Equivalent`, and will not be configurable,
but it will be possible to opt-out of the behavior by inspecting the
GVK of the admitted resource in the CEL expression.

xref:
https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules

#### Type safety

To keep failure policy easy to reason about, and to continue to use CEL in a
type-safe way we propose:

> - If a ValidatingAdmissionPolicy has a `spec.constraints` that matches a single

Reviewer 6: This seems like a pain for "pod-specy-things"

Author: epod-specy things are a tricky case. Alternatives that I've thought about:

1. Use a single configuration CRD, define multiple ValidatingAdmissionPolicy, one for each GVK containing the pod spec to be validated. Typesafe but requires lots of policy definitions.
2. Write a single CEL expression that can dynamically validates all the GVKs that contain a pod spec (CEL has no switch statement.. this gets ugly quick)
3. Introduce a capability to run CEL expressions on types nested within resources being admitted (e.g. `{scope: io.k8s.api.core.v1.PodTemplateSpec, expression: '...'`)

(1) can be made a bit more egronomic by grouping cel expressions by the type they validated so that a single policy could have validation expressions for multiple GVKs

Reviewer 7: What about just saying, if it matches a single type then you get type safety, and if it can handle multiple types you don't get type checking. Is that too much of a compromise?

Author: > if it matches a single type then you get type safety, and if it can handle multiple types you don't get type checking

We might end up here. I took a more careful look at all the requirements that impact type safety and I'm now starting to take the opinion that type checking should be informational-- we should emit warnings only.  See the updated "Type Checking" section.

Reviewer 3: WRT pod specy things... if the goal is to enforce policy, should we only enforce it against the pod itself?

Another issue with pod-specy things is that there is no guarantee that metadata will be identical between, say, the deployment and the pod it creates, and there is the possibility that a mutating webhook would bring a pod into compliance that doesn't touch deployments.

All of this means that trying to guess whether a higher-order resource would create a valid lower-order resource is brittle at best. This is why things like PSA don't strictly enforce policies against higher-order resources.

If we do want to enforce policies against a higher-order resource, a better approach may be to create a mock of the resource it is presumed to create, so we can have mutation webhooks, etc. apply.

I'm definitely not against multi-GVK policies, but it does present a scaling issue. Baking the operator behavior into each policy definition would make it expensive to accommodate changes to operator behavior.

Author: I've simplified this but retained exclude support.

  GVK, the CEL expression is allowed access to the full object in a typesafe

>   way. Otherwise, the CEL expression is allowed access to the metadata only.

Reviewer 4: > If a ValidatingAdmissionPolicy has a `spec.constraints` that matches a single GVK, the CEL expression is allowed access to the full object in a typesafe way. Otherwise, the CEL expression is allowed access to the metadata only

This won't be obvious to users to understand.

Author: I agree. We're planning to use dynamic typing so will become a non-issue.

- If there are any type checking errors (or if the CRD for the matched GVK does
  not exist):

>   - When a CRD is created/updated: The type check errors are detected by an

Reviewer 3: What is the impact if a CRD is updated such that pre-existing policies are put in an error state? Are requests denied?

Author: I think that's the right thing to do.  If the CRD updated intended to make the policy more restrictive (but accidentally broke the policy) then denying requests is less risky than allowing them.  And if the API server were restarted, it can become impossible to know what the previous CRD state was. I'll clarify.

Author: One clarification here: All requests fail, but they are only denied if the `FailurePolicy` is set to `Fail`.

    control loop watching the CRDs with an informer in the API server, and
    reported in the status of the ValidatingAdmissionPolicy.
  - When a policy is validated: runtime evaluations of the expression are
    skipped, and the validation is reported as failed, with a type check error.

> The main disadvantage of this approach is that CEL rules that would have

Reviewer 3: There is also the concern of version skew:

If k8s version N + 1 introduces a new Pod field, and a policy definition gets rewritten to accommodate that new field, then older versions of K8s cannot use that updated policy definition, even if it has other fixes, and even if the definition is written in such a way as to gracefully handle cases where the expected field is undefined.

Author: Dynamic code could guard access to fields with CEL `has(x.fieldname)` checks. This is a big ask for policy authors ("check every field you access to see what Kubernetes version it was introduced as and if that version is newer than the oldest version your policy supports, figure out how to write the policy so that it still works without that field..."). Granted, if the policy author does get this right, they get the benefit.  But for everyone else, they just loose type safety.

Reviewer 3: To make this a bit more concrete, here is an example of a constraint template looking at seccomp policies that addresses the transition from using the annotation to using a field to define them:

https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/pod-security-policy/seccomp/template.yaml

Author: Thanks! I've added a use case and linked to this example.

    # metadata is always accessible
  - rule: "object.name.startsWith('xyz')"
```

Alternative Considered: Do not use CEL type-checking and instead dynamically evaluate
all CEL expressions.

Pros:

- Can write a single CEL expression that processes multiple GVKs

Cons:

- All errors are reported during admission control validation. If instead
  type-checking is used, CEL type-check errors can be surfaced much earlier in
  the status of `ValidatingAdmissionPolicy` resources when the
  `ValidatingAdmissionPolicy`, configuration CRDs and configuration resources
  are written to the API Server. This provides a workflow where type-checking
  can be used to detect errors early, preventing them from occurring during
  admission control validation and making it safer to operate a policy using
  `failureMode: Fail`.

#### Failure Policy

For in-process validation there is no remote request, so errors should be
deterministic. We believe this significantly reduces the risk of "fail closed"
admission control as compared to webhooks.

Because failure policy is most often selected based on the need to guarantee

> enforcement, we propose defaulting failure policy to "fail" and allowing it to

Reviewer 4: Would it make sense to default to fail open until we have a way to ensure a policy isn't going to break the cluster? e.g. https://github.com/kubernetes/kubernetes/issues/9215

Author: I've gone back and forth on this one.  One one hand, defaulting to fail open can make rollouts safer, one the other hand, it's a problem for compliance where defaulting to fail closed is very much needed.  Webhooks already default to fail closed, so I'm tempted to take that path purely as an API consistency choice-- we are claiming that fail closed is less dangerous with CEL.. Curious what others think.

Author: This is covered in the match criteria section.  We intend to support exclusion on each major match field. How we do this varies by field. For example scope has only two options (cluster, namespaced) so we don't need a exclude for that field, just  use `scope: [Namespaced]`, but for unbounded fields there needs to be an explicit exclude capability.

be configured on a per-rule basis:

```yaml
  validations:
    - rule: "object.spec.xyz == configuration.x"
      failurePolicy: Ignore # The default is "Fail"
```

TODO: Metric for ignored failures?

#### Safety measures

To prevent clusters from being put into a unustable state that cannot be
recoverd from via the API, admission webhooks are not allowed to match
`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.

We will extend this approach:

- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
  `MutatingWebhookConfiguration`.

Note that this does allow `ValidatingAdmissionPolicy` to match
`ValidatingWebhookConfiguration`.

> We are also considering not allowing a `ValidatingAdmissionPolicy` to match the

Reviewer 6: While providing tight protection, this makes hooking CRDs a pain.  I think I'd be biased towards allowing this.

Author: I'll drop this note since it's possible to self-rescue if a mistake is made for something like this.

scoped. We can support namespace scoped policy configuration as follows:

- If the configuration resource is namespace scoped, it implicitly matches
  resources only in the namespace it is in, but may further constrain what
  resources it matches with additional match criteria.

Benefits: Allows policy of a namespace to be controlled from within the
namespace. 

#### CEL expression scoping

CRD validation rules are scoped to the schema at the location in the OpenAPIv3
where they are defined. This make validation rules far easier to author by
eliminating the need to dereference from the root of an object to the field that
needs to be validated. Should we provide something similar?

Use cases:

- Dereferencing to deeply nested fields in CEL. A `scope` field containing a
  field path be sufficient. E.g. `spec.containers[*].image`
- More ambitiously, some way to scope a CEL expression to a type found nested in
  multiple kinds, e.g. "io.k8s.api.core.v1.PodTemplateSpec" or
  "io.k8s.api.core.v1.Container" would help policy authors apply policies more
  broadly and uniformly (e.g. to match initContainers, containers,
  ephemeralContainers).

> #### Reporting violations

Reviewer 4: Don't we need this for phase 1 to return some message back to the user?

Author: Yes, I've moved it up and made an attempt to provide more detail on what the exact API will be.

We need:

- Formatted messages (need support for string formatting. Allow a CEL expression
  to be used to format the whole string? Or allow CEL expressions to evaluate
  template values?). Note that CRD Validation Rules do NOT support formatted
  messages (but if we add support for formatted message for admission control,
  we should consider adding the feature to CRD validation as well).
- severity (e.g. "Warning")
- Audit Annotations - Information recorded to the audit system but not reported
  back in the response.
- status types (e.g. Forbidden can be returned from a webhook using
  HTTP 403) (can a match rule override this? or can it be selected via a CEL
  expression?)

> - Similar to how each match has configuration, can each match also select

Reviewer 3: Policy frameworks provide other reporting mechanisms, such as metrics.

For a user who consumes this KEP (lets assume via a policy framework), but also has policies enforced via the framework's webhook (e.g. b/c the framework supports a functionality this KEP doesn't), is there a way to unify how violations are reported?

If not, it will be hard for a policy framework to leverage this KEP transparently.

Author: This is very important both for use cases that migrate from webhooks to this KEP and for use cases that will be using both indefinitely.

We are planning to support all the same admission review response fields that a webhook may return. We need to provide sufficient control of how messages are formatted so that a message previously being formatted in the code a webhook can be re-implemented using this KEP.

I don't yet know what will be possible for metrics.  At what granularity are the metrics for OPA gatekeeper?

Reviewer 4: https://open-policy-agent.github.io/gatekeeper/website/docs/metrics has a list of metrics provided by Gatekeeper, specifically violations from the webhook and audit.

Author: Thank you! I've added a short section about metrics that briefly summarizes what we should track, based on both what admission webhook metrics provide and what OPA gatekeeper provides.  I've include the link you provided.

  the severity of messages? It make it possible to configure a policy as
  warning only when first enabling it. Some policy frameworks support this.

#### kube-apiserver authorizer checks

kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
as a way of doing things like:

- Validate that only a user with a specific permission can set an enum to the
  "HOLD" value.
- Validate that only a controller responsible for a finalizer can remove it from
  the finalizers field.
 
 Concerns: "Is joe authorized to do this"? That only works for the objects joe
 creates, but not objects that get created on joe's behalf by a controller.

>  Ditto for updates. I heard someone cite PSP as an example for why it's needed,

Reviewer 6: Also used by
1. PodSecurityPolicy (kube)
2. CertificateApproval (kube)
3. CertificateSigning (kube)
4. OwnerReferencesPermissionEnforcement (kube)
5. network.openshift.io/ExternalIPRanger
6. route.openshift.io/IngressAdmission
7. scheduling.openshift.io/PodNodeConstraints
8. network.openshift.io/RestrictedEndpointsAdmission
9. security.openshift.io/SecurityContextConstraint
10. security.openshift.io/SCCExecRestrictions

and that's just the in-tree in openshift, doesn't include other project like kubevirt and open-cluster-management. The pattern is commonly and successfully used both in the kube project (outside of PSP) and by admission consumers elsewhere.

Problems in PSP were multi-faceted and secondary authoriation checks will be required to replace a significant portion of admission plugins.

Reviewer: Even if it's not appropriate for all scenarios, I think admission that reframes some attribute of the incoming object into an authorization check would be pretty straightforward to implement in-process and as a webhook, would be pretty simple to express in a cel expression, and would be pretty powerful.

I think it was uniquely problematic for PSP for two reasons:
* PSP guarded pods, which are almost completely created by controllers on behalf of other users. That's not the case for most objects, so there's plenty of object types where authorization checks of an attribute of the object against the requesting user could make sense.
* PSP tried to compensate for the controller-created scenario by _also_ checking the pod's service account authorization, which was _really_ confusing.

Reviewer 7: What are the advantages to using authz over a custom configuration object for these examples?

Another disadvantage to using authz is it means the admission execution is no longer deterministic, and we can't take advantage of the performance analysis for validation.

Author: Per offline discussion, we're going to accumulate more use cases and start sketching out a proposal. I'll make some general notes in the KEP in the short term.

 but IMO that was an anti-pattern of PSP, and one that we explicitly decided to
 omit from PSA

> #### Access to namespace metadata

Reviewer 6: note to self, didn't make it past here.

more control surfaces, and the delay in delivering these control surfaces due to
K8s' rollout period.

Pod Security Admission is a similar solution, but does not attempt to duplicate
the control granularity that PSP provided.

There are numerous in-tree embedded controllers.

The existence of security regimes like the CIS Kubernetes Benchmarks highlight
the values of standardized controls. Automating their enforcement, where
possible, will make it easier for users to lock down their clusters.

With the advent of CRDs, and the drive to make the resources they define
first-class entities, the footprint of Kubernetes extensions is set to grow for
the foreseeable future. This KEP allows authors of such extensions to provide
policy primitives similar to PSP or PSA, putting them on equal footing with
in-tree functionality.

With the reduced infrastructure footprint and demonstrated demand for a
customizable, built-in mechanism for extensible policy, this KEP fills a
community need. It is not intended to replace validating admission webhooks
altogether, however, since these can support functionality that may not make
sense to provide in-tree.

> ### Goals

Reviewer 6: sig-auth has previous talked about trying to find a way to restrict access from a daemonset pod to a customresource/foo that has Foo.spec.NodeName set to the Node.metadata.name of the pod bound to the particular SA token. This is tantalizingly close because user.Extra contains `authentication.kubernetes.io/pod-uid` to locate a pod, determine a Pod.spec.NodeName.

A built-in that does that may be well received and unlock many use-cases.  Exploring the idea may be useful.  If most also require controlled read permission, then its probably better to create something specifically for the purpose.

Author: I've added a note to the KEP. Sounds like we will need to investigate. Looking up the pod (or any other additional resources) is not something we were planning to allow, but I'd still like to understand this one better. I'll bring it up with sig-auth.

can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  config:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  match:
    rules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
      scope:       "*"
  validations:
    - expression: "object.spec.replicas <= config.maxReplicas"
      reason: Invalid
      messageExpression: "'object.spec.replicas must be no greater than ' + config.maxReplicas"
      # ...other rule related fields here...
```

> The `spec.validations[*].expression` fields contain CEL expressions. If a

Reviewer 6: ```suggestion
The `spec.validations[*].expression` fields contain CEL expressions. If an
```

- 'config' - configuration data of the policy configuration being validated 

See below "Reporting violations to Clients" for more detail about how violations
are reported.

##### Policy Configurations

Policy configurations are the interface between `ValidatingAdmissionPolicy`
authors and the administrators that configure the policies for clusters. The
configurations allows policy definition authors to define how exactly a
`ValidatingAdmissionPolicy` may be configured using a OpenAPIv3 structural
schema.

How these configuration CRDs are defined has not yet been decided.
Alternatives are summarized here and discussed in more detail below:

<<[UNRESOLVED jpbetz, liggitt, ?? ]>>
Reach consensus on this.
<<[/UNRESOLVED]>>

| Design Alternative                                | Summary                        |
| ------------------------------------------------- | ------------------------------ |
| `PolicyConfiguration` kind with config reference  | Match criteria is defined in builtin type which references a config  | 
| OpenAPIv3 `$ref` in CRDs                          | Match criteria is defined by k8s as a OpenAPIv3 type and CRDs schemas are modified to be allowed to reference it |

> | Duck Typed CRDs                                   | Require CRDs be defined with match field of correct structure |

Reviewer 6: My initial impression is that I like this best.

Author: This remains my 1st choice as well. The main issue with this one is that the `spec.match` can be malformed (both in CRD schemas and in the data of CRs). We previewed this one with API review and got feedback that we should explore alternatives. I think I've now got a extensive list, and "`PolicyConfiguration` kind with config reference" is the only other viable contender from what I can tell (all other alternatives have _major_ problems).

I'll update this alternative with more detail about what the possible failure cases are and how they could be handled.

This is the alternative shown in the initial examples of this KEP.

The idea of this alternative is to use duck typing: If a CRD author includes a
`spec.match` schema in their CRD conforming to the schema expected (described
below) and then for the apiserver access the data and treats it as match
criteria.

If the schema for `spec.match` it is malformed, errors are reported in the
status of the `ValidatingAdmissionPolicy`.

A partial `spec.match` schema (subset of the full schema) is okay so long as
only optional fields are omitted. But any unrecognized field in the `spec.match`
would not be allowed. 

Pros:

- A single resource is used to configure both the match criteria and
  configuration params of a policy.
- Policy authors can choose to only provide the subset of `spec.match` schema
  that apply to the configuration of the policy they have defined. E.g. a policy
  that is already constrained to pods by the policy definition does not
  need to expose the match criteria for GVR matching.

Cons:

- API server must check for a wide range of error conditions and define how
  exactly it handles each of them.
- The way CRDs expose Match criteria is entirely implicit.

> - If the `spec.match` schema is incorrectly defined, CRD author might not

Reviewer 6: What if you allowed an annotation on a CRD manifest that meant: "inject the correct .spec.match during admission and keep it up to date with a controller"?

Author: Oh nice. I've added this to the alternative proposal.

Audit. Match criteria will also use ordered list of rules similar to these other
systems.

In order for policy configuration resources to declare match criteria, the
corresponding configuration CRD schema must have a `spec.match` property. This
property must conform to the below "matching schema template". This ensures that
the match criteria is in the format that API server expects (the API server will
be using duck typing here since there is no established way to do polymorphism
across CRDs). The schemas of these fields in the configuration CRD may omit any
optional properties; policy definition authors should only include the parts of
the "match schema template" that are useful for configuring a particular policy.

Proposed matching mechanism:

- Ordered list of matches. First match that is true determine if the request is
  validated by the policy-- if exclude is false, it is validated, otherwise it
  is not.
- Supported criteria for each match:
  - GVR (resource apiGroup + resource)
  - scope (cluster|namespace) 
  - operation (HTTP verb) 
  - exclude (if criteria matches, match result is "no match")
  - namespace selectors
  - object selectors
  - resource name

>   - user/userGroup

Reviewer 6: user.Extra?  (pod claims are here)

also, this is as far as I got.

Author: Added.

We have divided this proposal into phases, all of which must be completed before
this feature graduates to beta. Our goal is to size the phases so that each
can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  config:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  match:
    - resourceRules:
      - apiGroups:   ["apps"]
        apiVersions: ["v1"]
        operations:  ["CREATE", "UPDATE"]
        resources:   ["deployments"]
  validations:
    - expression: "object.spec.replicas <= config.maxReplicas"
      reason: Invalid

>       messageExpression: "'object.spec.replicas must be no greater than ' + config.maxReplicas"

Reviewer 8: nit related to [this comment](https://github.com/kubernetes/enhancements/pull/3492/files#r977164813) - maybe this should be `messageExpression: "'object.spec.replicas must be no greater than ' + string(config.maxReplicas)"` instead? 
I tried making it a CRD validation rule - `size('object.spec.replicas must be no greater than ' + self.maxReplicas) < 75` so it'd evaluate to a bool - and got `ERROR: <input>:1:54: found no matching overload for '_+_' applied to '(string, int)'`.

Author: Good find. I'll update the example. This does make message formatting a bit less pleasant, which makes me wonder if we'd be better off just offering a simple template format like "object.spec.replicas must be no greater than {{config.maxReplicas}}" where only dereferencing is supported.

Reviewer 8: Were you thinking of using [text/template](https://pkg.go.dev/text/template)? The syntax reminded me of it, and since it's in the standard library I don't think we'd need to do much work to add it in. It's rather complex/Turing-complete (would be kinda funny to have error message templates be Turing-complete while the actual policy rules weren't, heh), so maybe using it would be a bit too much.

We could roll our own with a similar syntax, but there's a couple things to think about:
* Array subscripting - would be simple for users (`{{ config.someArray[2] }}`) but we'd have to parse it/write support for it.
* Type checking - we'd have to implement some kind of system even if we want just informational type checking. As an aside, as far as I know `text/template` lacks a type system for what that's worth.

We could also try for a simple string-replace system, though I don't see how we could do subscripts/type-checking that way, so I don't really think I'm a fan of that route, at least on paper.

I agree that having to potentially write `string(someVar)` would make CEL-based message expressions a little unwieldly, though at the same time I'm a little concerned about the potential complexity other options would involve.

Author: I'm comfortable ruling out text/template. If we go for simpler string formatting, json path is probably the least risky choice since CRDs already use them for a few things. Right now I'm tempted to just stick with CEL.

@TristonianJones any recommendations for string formatting with CEL?

Author: Oh, one other thought here, we could do either:

`object.spec.replicas must be no greater than {{<cel expression>}}`

Or,

```yaml
messageFormat: "object.spec.replicas must be no greater than {0}"
messageArgs: ["<cel expression>"]
```

This eliminates the need to do the type casts to string.

Reviewer 8: I like embedding CEL expressions with `{{ }}`. Since it also prevents the need to nest quotes (`" ' ' "`), maybe we could end up rolling the `messageExpression` field into `message`, since for error messages without CEL expressions they'd end up being equivalent?

Author: Embedding `{{<cel expression}}` in strings does mean that all CEL expressions must have escaping rules defined to handle `{{` or `}}` inside the CEL expression. If we use `message` and `messageArgs` we don't have to deal with this, but it makes the simple cases a bit more verbose to write.

Reviewer 9: @jpbetz @DangerOnTheRanger Is the primary issue that making `message` a CEL expression would require type-casting of parameters? There's always the option of introducing a `string.format("message {1}, {2}", [args])` method into CEL. I'm not sure you need to invent a new syntax around it in the YAML.

| user.Extra                               | (in WH AdmissionReview)      | phase 2 see "Secondary Authz" section  |
| permissions (RBAC verb)                  | RBAC                         | phase 2 see "Secondary Authz" section  |

WH = Admission webhooks, P&F = Priority and Fairness

Match criteria must be declared in the `spec.match` field of policy
configuration resources (see `ReplicaLimit` in the above example) and will be
declared with API types in a format similar to admission webhooks, P&F, RBAC and
Audit.

Proposed matching mechanism:

- Matching criteria is defined as an ordered list of match objects (similar to
  [Audit
  Policy](https://kubernetes.io/docs/reference/config-api/apiserver-audit.v1/#audit-k8s-io-v1-Policy))
  - If a match object evaluates to true
    - If the match object is an "exclude", skip validation
    - Otherwise perform validation
  - If all match objects in the list evaluate to false, skip validation
- The criteria of each match object in the list is applied conjunctively (ANDed)

<<[UNRESOLVED, maxsmythe, ?? ]>>
Consider an alternative where instead of a list of match criteria, we have a

> single match criteria where it is possible to handle excludes. I.e. we sanely

Author: Yes, phase 1.  Matching is one of the least nailed down phase 1 features. I've added an example and some rationale. @maxsmythe had offered to help refine this part of the proposal, so we'll be doing additional design work here over the next week or so.

generalize how labelSelectors support excludes to all the other match criteria?
<<[/UNRESOLVED]>>

Examples:

```yaml
# Matches resources with {namespace a, label x=1} and {namespace b, label x=2}
# but not {namespace a, label x=2} or {namespace b, label x=1}.
match:
- namespaces: ["a"]
  labelSelector:
  - key: x
    operator: In
    values: ["1"] 
- namespaces: ["b"]
  labelSelector:
- key: x
    operator: In
    values: ["2"] 
```

```yaml
# Matches {namespace a, label x=2}, {namespace b, label x=1}, {namespace b, label x=2}
# but not {namespace a, label x=1}.
match:
- namespaces: ["a"]
  labelSelector:
  - key: x
    operator: In
    values: ["1"] 
  effect: Exclude
- namespaces: ["a", "b"]
  labelSelector:
  - key: x
    operator: In
    values: ["1", "2"] 
```

<<[UNRESOLVED jpbetz, maxsmythe, ?? ]>>

> Provide wildcard prefix support to namespaces, resource name or any other fields?

Author: Thx. For this 1st alpha, I'm keep things simple drop any prefix/postfix (globbing) match support in this proposal. CEL is a good escape hatch for anyone wanting more expressitivity.

Or better to defer this type of check for CEL expression evaluation?
<<[/UNRESOLVED]>>

<<[UNRESOLVED jpbetz, maxsmythe ]>>

> Offer both GVR and GVK matching?

Author: I don't have any compelling use cases and so I'll drop this. Matching scale can be handled via "*/scale" as a resource so it doesn't require kind matching, and the explicit wildcard has nice readability properties when expressed as a resource.

@maxsmythe I did notice that OPA gatekeeper matched on GVK instead of GVR, so this may impact future integration. Maintaining a mapping outside of k8s between GVK/GVR seems like trouble, so not sure what options to recommend but maybe we can discuss more separately.

  CEL in the future.

###### Exclude matching

Benefits:

- Exclude rules make it easier for policy definition authors to do things like
  exclude kube-system or other special purpose namespaces from policies that apply
  to all "normal" namespaces.
- Default rules make it easier to apply principal of least privilege to
  policies. I.e. ensure that newly created resources default to a least
  privilege state until configured otherwise.

Exclude rules can be modeled after the approach exstablished in Audit where
level=None on a match rule can be used to exlude any rules beneath it. E.g.
`effect: Exclude`.

If we want this, it is essential that we introduce it early, the exclude inverts
the semantics of a match, adding it in a later release is _very_ incompatible.

#### Reporting violations to Clients

This section focuses on how information is reported back to clients in
when validations fail.

<<[UNRESOLVED ]>>
Lots of open questions about this part of the KEP that need resolution:

- Should enforcement be configured on the policy definition, the policy
  configuration, or both (with some sort of override mechanism)?

> - Are Deny, Warn and Audit the right set of enforcements?

Author: I agree. I've added a note.

> - Should there also be an Ignore enforcement level (disablement option)?

Author: Cleaned this up.

- Recording audit annotations requires an audit key. How to handle? E.g. if a
  policy defines multiple validations and the configuration is set to
  enforcement=Audit, what audit keys does the policy use?
- If policy definitions can set enforcement, should it be per-validation
  expression or per-policy?
- If policy configurations can configure enforcement, is there only a single
  field with a single enforcement option that can be set, or are there more fine
  grained controls for handling each of the validations that can fail in a
  policy?
<<[/UNRESOLVED]>>

Goals:

- Feature parity with `AdmissionReview`
  - Support allow/deny result, warnings, audit annotations
  - Support for reasons/codes
- Ability to format message strings

High level proposal:

- Each policy may set a `denyReason` and/or `denyCode`
- Each validation may define a message:
  - `message` for plain strings

>   - `messageExpression` for a CEL expression evaluates to a string

Reviewer 8: Is `messageExpression` not type-checked/informationally type-checked the way policy validation rules are? I think there could be situations where trying to eval `messageExpression` fails at runtime. I don't think there's generally too much we could do about it other than emit some kind of "messageExpression evaluation error" message, but it might be worth noting that it could happen.

There could also be potential edge cases where a policy rule validation fails and whatever causes the validation failure also causes the corresponding `messageExpression` to fail to evaluate properly, as opposed to `messageExpression` being invalid no matter what the input is, and that might prove tricky to debug. For example, say we have a rule along the lines of `someProp == "string 1" || someProp == "string 2"` and then have a `messageExpression: "invalid value for someProp:" + someProp`.
If `someProp` does not contain a string for some reason, then it'll result in a runtime error while trying to evaluate `messageExpression` since as far as I know CEL is relatively strict about concatenating. `self + 2 != "some string"` as a CRD validation rule defined on a string property gives me `ERROR: <input>:1:6: found no matching overload for '_+_' applied to '(string, int)'`, for instance.

Those edge cases could be mitigated by wrapping every variable used as part of `matchExpression` in `string()`, i.e, `messageExpression: "invalid value for someProp:" + string(someProp)`, but since that's not always strictly necessary depending on the type of `someProp` it could be a mildly subtle gotcha. We could always have the CEL linter catch that sort of thing, too.

Author: Yes, this is all a bit messy when things go wrong.

I was imagining that `messageExpression` would receive treatment almost identical to the `expression` except that any errors (including runtime cost limits) would result in fallback to a auto-generated message, e.g. "expression spec.x != spec.y" and the error from the messageExpression would be logged but not surfaced to the user. I wonder if there is anything better we can do than that?

`messageExpression` should also have a lower runtime cost limit than `expression`. I hadn't thought about that until now.

+1 for linting

Reviewer 8: For auto-generated messages, maybe it would be nice to know what term(s) of the policy rule expression evaluated to false? That could come in handy for debugging. For instance, some kind of error message like:
```
someProp != "badvalue1" && someProp != "badValue2" && someProp != "badValue3"
___________________________^^^^^^^^^^^^^^^^^^^^^^^___________________________
```
It ought to save a few minutes even for simple policy rules, and some headache for more complicated rules where it's not immediately obvious what failed and why. That said, I'm not sure cel-go has the execution tracing that we'd need to implement that kind of thing at the moment, unfortunately.

One simple alternative would be to also print the arguments passed to policy rule that failed. There still might potentially be a little bit of debugging work that users would have to do to figure out what term of the rule failed, but it would be very simple to implement.

- allow/deny, warnings, audit annotations will be collectively referred to as
  `enforcement`, which will have the following levels:
  - `Deny` (the default),
  - `Warn`
  - `Audit` (must also set `auditKey` field if used)
  - `Ignore` (useful for disabling a validation without deleting it)
- Policy definitions set enforcement level per-validation

> - Policy configurations can also set an enforcement level that overrides all

Author: I've dropped this in favor of having the policy configurations set all enforcement levels.

dynamically. Informational type checking will be provided (except for aggregated
API server types), but will be surfaced only as warnings. See "Alternatives
Considered" section for details of all the alternatives we reviewed when
selecting this approach.

Type checking is still performed for all expressions where a GVK can be matched
to type check against, resulting in warnings, e.g.:

```yaml
...
status:
  expressionWarnings:
    - expression: "object.foo"
      warning: "no such field 'foo'"
```

#### Failure Policy

For in-process validation there is no remote request, so errors should be
deterministic. We believe this significantly reduces the risk of "fail closed"
admission control as compared to webhooks.

Because failure policy is most often selected based on the need to guarantee
enforcement, we propose defaulting failure policy to "fail" and allowing it to
be configured on a per-policy basis:

```yaml
  failurePolicy: Ignore # The default is "Fail"
  validations:
    - expression: "object.spec.xyz == configuration.x"
      
```

<<[UNRESOLVED jpbetz ]>>
Should failurePolicy be per-validation expression or per-policy?
<<[/UNRESOLVED]>>

<<[UNRESOLVED jpbetz ]>>

> Also allow for a `failurePolicy` field on policy configuration resources?

Author: SGTM, dropping it.

configuration to intercept/guard writes to admission configuration while
preventing deadlock - Add feature to configure a set of webhooks to intercept
other webhooks https://github.com/kubernetes/kubernetes/issues/101794.

Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
higher level. This could be added later.

#### Singleton policies

<<[UNRESOLVED jpbetz, ?? ]>>
Should this be supported in phase 1? Can it be deferred to phase 2 without
necessitating breaking API changes? What are the implications on how violations
are reported.
<<[/UNRESOLVED]>>

For a policy that requires no configuration, we would prefer not to ask users
create configuration CRD and configuration resource that serve no actual
purpose. Instead, we will allow a `ValidatingAdmissionPolicy` to define
a singleton policy, the recipe is:

- Set a `policyType: Singleton` field
- Exclude `spec.config`

For example:

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:

>   policyType: Singleton

Reviewer 6: Why does it need a different policy type instead of simply not specifiying config field?

  match:
    rules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
      scope:       "*"
  validations:
  - expression: "object.spec.replicas < 100"
```

Adding a `policyType` field makes it easier for a user examining the policy to
observe that the policy is different that normal (configurable) policies. This

> is much less obvious if we infer that a policy is a singleton by the state of

Reviewer 6: not so sure. If I don't fill in config and I don't reference any config, it seems pretty straightforward.

the other fields. It also makes validation errors reported when validating a
`ValidatingAdmissionPolicy` much easier to communicate to the policy author.

### Phase 2

All these capabilities are required before Beta, but will not be implemented
in the first alpha release of this enhancement due to the size and complexity
of this enhancement.

#### Namespace scoped policy configuration

For phase 1, policy configuration resources were only allowed to be cluster
scoped. We can support namespace scoped policy configuration as follows:

- If the configuration resource is namespace scoped, it implicitly matches
  resources only in the namespace it is in, but may further constrain what
  resources it matches with additional match criteria.

> Benefits: Allows policy of a namespace to be controlled from within the

Reviewer 6: perhaps add a "as an example, ResourceQuota works this way"

namespace. 

#### CEL expression scoping

<<[UNRESOLVED jpbetz, ?? ]>>
This is not yet well defined or agreed on. Reach agreement or drop.
<<[/UNRESOLVED]>>

CRD validation rules are scoped to the schema at the location in the OpenAPIv3
where they are defined. This make validation rules far easier to author by
eliminating the need to dereference from the root of an object to the field that
needs to be validated. Should we provide something similar?

Use cases:

> - Dereferencing to deeply nested fields in CEL. A `scope` field containing a

Reviewer 6: At this moment, I have no strong feelings either way about this feature.

For example, to validate all containers:

```yaml
  validations:
    - expression: "self.name.startsWith('xyz-')"
      scopes: ["spec.containers[*]"]
```

This could be extended to validate all kinds of containers in a pod if scopes
supports a homogeniously typed list:

```yaml
  validations:
    - expression: "self.name.startsWith('xyz-')"
      scopes: ["spec.containers[*]", "initContainers[*]", "spec.ephemeralContainers[*]"]
```

`ephemeralContainers` might not exist due to version skew (it's a relatively new
field), in which case that scope would be dropped (but a warning would be added
to the status).

#### Secondary Authz

<<[UNRESOLVED jpbetz, deads2k, tallclair, ?? ]>>
We have general agreement to include this as a feature, but need to provide a
more concrete design.
<<[/UNRESOLVED]>>

kube-apiserver authorizer checks (aka Secondary-authz checks) have been proposed
as a way of doing things like:

- Validate that only a user with a specific permission can set a particular
  field.
- Validate that only a controller responsible for a finalizer can remove it from
  the finalizers field.

This could be supported by matching criteria, or via CEL expression access, or both.
 

> Use cases:

Reviewer 6: not to self, I still owe you an answer about whether we only check self or whether we choose identities based on a field.  I think I remember seeing identify chosen by field in the open-cluster-management enhancements repo.

- OwnerReferencesPermissionEnforcement (kube)
- network.openshift.io/ExternalIPRanger
- route.openshift.io/IngressAdmission
- scheduling.openshift.io/PodNodeConstraints
- network.openshift.io/RestrictedEndpointsAdmission
- security.openshift.io/SecurityContextConstraint
- security.openshift.io/SCCExecRestrictions

From deads2k:

> Note that user.Extra in AdmissionReview has pod claims, which are valuable.

> sig-auth has previous talked about trying to find a way to restrict access
> from a daemonset pod to a customresource/foo that has Foo.spec.NodeName set to
> the Node.metadata.name of the pod bound to the particular SA token. This is
> tantalizingly close because user.Extra contains
> authentication.kubernetes.io/pod-uid to locate a pod, determine a
> Pod.spec.NodeName.

> A built-in that does that may be well received and unlock many use-cases.
> Exploring the idea may be useful. If most also require controlled read
> permission, then its probably better to create something specifically for the
> purpose.

Looking up the pod (or any other additional resources) is not something we are
currently planning to support in this KEP, but the use case is interesting and

> we should investigate with sig-auth.

Reviewer 6: Agreed.  This was a CVE for numerous distributions.

Reviewer 4: +1

#### Default configurations

<<[UNRESOLVED jpbetz, maxsmythe ]>>
Should we attempt to support default configurations? We do not have viable
design proposal yet, despite considering quite a few options.
<<[/UNRESOLVED]>>

Idea: Allow policy definitions to provide a default configuration that is
enforced for a resource only if no other policy configurations match that
resource (but the policy definition match rules do match the resource).

Default configurations are to support mechanically. Options considered:

> - Embed the default configuration into the policy definition. Messy to

Reviewer 6: This is interesting.  Perhaps a default policyrule that isn't allowed to reference any configuration?

  implement, less appropriate for use cases where the default configuration is
  intended to be configurable on a per cluster basis since it requires handing
  out the same permissions required to modify the policy definition.
- Policy definition references the default configuration (`defaultConfig: "<configurationResourceName>"`).
  Same permission problem as previous option.
- For cluster scoped policy configuration objects, use a singlton configuration
  resource e.g. "<name-of-policy-definition>-default". Can be combined with
  `defaultConfiguration: true` in place of normal matching criteria to make it
  obvious that resource is a default configuration.
- For namespace scoped policy configuration objects, a singleton doesn't work
  (unless there is a good namespace to put it in that would work for all
  clusters).
  - To always have the default configuration cluster scoped, policy defintions
    could be allowed to reference two configuration CRDs, one for cluster scope
    and another for namespace scope. Having two CRDs is VERY undesirable.

None of these options seem satisfactory to me.

#### Access to namespace metadata

> - Namespace labels and annotations are the most commonly needed fields not

Reviewer 6: Can I get an UNRESOLVED, on this?  I don't have a firm opinion at the moment.

#### Safety Features

<<[UNRESOLVED jpbetz, ?? ]>>
Decide exactly what we are willing to commit to here.
<<[/UNRESOLVED]>>

- Configurable admission blocking write requests made internally in
  kube-apiserver during server startup (like RBAC default policy reconciliation)
  making it impossible for a server to start up healthy. (This is not specific to CEL?)
- Ability to skip specific resource types - Admission Controller Webhook
  configuration rule cannot exclude specific resources:
  https://github.com/kubernetes/kubernetes/issues/92157

#### Aggregated API servers

Main complications (provided by @liggitt):

- The API server validating/persisting the ValidatingAdmissionPolicy
  instances isn't the same one serving aggregated types, so wouldn't necessarily
  have schema info to check type safety.

- The aggregated API server is responsible for enforcing admission on its custom
  types, so the implementation that reads ValidatingAdmissionPolicy
  instances and enforces them would have to live in k8s.io/apiserver and be
  active in aggregated API servers to enforce admission on aggregated types
  effectively (same as admission webhooks today).

> Plan:

Reviewer 6: This plan looks reasonable to me.

- Do not offer type checking for aggregated types.
- Support ValidatingAdmissionPolicy in aggregated API servers.

#### Metrics

> Goals:

Reviewer 6: At what granularity do you need latency metrics?

Author: I'm in favor of a metric per {policy, validation expression}

The next level of granularity would be {policy, configuration, validation expression}, but that the number of configurations can become quite large, so I'm hesitant to go that far.

I've put this in the KEP.

Reviewer 3: Are we concerned that policy/expression names in cleartext/unacld metrics might leak policy surface?

      resources:   ["deployments"]
      scope:       "*"
  validations:
  - expression: "object.spec.replicas < 100"
```

#### Use Case: Shared Configuration

User wishes to define a configuration for a list of banned words that may not be
used in any of a wide range of identifiers in the cluster (resource names,
container names, ...).

- Configuration CRD is defined to hold the list of banned words.
- Multiple policies are defined for different resources. The policies all
  reference the same configuration CRD. 
- Policy must be able to specify it's own matching rules since the configuration
  applies cluster wide.
- A single custom resource is defined with the list of banned words (but has not
  matching rules of its own).

```yaml
apiVersion: rules.example.com/v1
kind: BannedWords
metadata:
  name: "banned-words.example.com"
spec:
  bannedWords:
  - glitter
  - rainbow
```

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "policy1.example.com"
spec:
  match:
    rules:
    - apiGroups:   ["*"]
      apiVersions: ["*"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["pods"]
      scope:       "*"
  validations:

>   - rule: "!object.name in config.bannedWords"

Reviewer 8: Do we need to consider resource constraints/failure modes for (potentially) relatively resource-heavy CEL expressions like this one?  At a high level, I feel like what was done for [CRD validation rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#resource-constraints) would be applicable (estimate how expensive the rules are to run before we create/apply and fail if it's too much).

I think we could have:
* Failure due to an individual validation rule exceeding a per-CEL-rule estimated cost limit
* Failure due to the sum total of a `ValidatingAdmissionPolicy`'s rules exceeding the per-policy estimated cost limit (if we decided to have one - I think it makes sense for the same reasons as CRD validation rules, to cap resource usage on a single policy)
* Failure due to a request timeout while evaluating a rule
 
That said, cost estimation [relies on type checking](https://github.com/google/cel-go/blob/1b994e0945736cb8ef2b51aedf4b30fdeeb7bc1c/checker/cost.go#L284), so maybe that'd mean cost limits would only be available some of the time if policy rules aren't always type-checked.

Author: Thanks for bringing this up.

I agree that estimated cost is, unfortunately, not something we can offer with any kind of guarantees attached due to the already listed issues that have prevented use from enforcing type safety. We could compute it for the same cases where we provide informational type checking, in which case we could report any cost limit violations in the same way we report type checking violations. To make matters worse, we will be handling both built-in and CRD types, and built-ins do not have their `max{Length,Properties,Items}` limits documented in schemas, which further limits estimated cost usefulness. I'll make note of this in the KEP.

Runtime cost limits are something that we should do, as is a context cancellation backstop. Both have `FailurePolicy` implications. I'll add this into phase 2.

Author: Shift-left checks for CEL estimated cost would be really helpful here. Something to think about when working on the CEL expression linter?

  - rule: "!object.spec.initContainers.any(c, c.name in config.bannedNames)"
```

Similar Use Case: A cluster administrator wishes to use a single policy
configuration to manage a network policy that must be enforced across multiple
Kubernetes kinds that contain relevant networking fields. It is possible to
implement by having multiple `ValidatingAdmissionPolicy` resources that all
reference the same `spec.config` CRD but that each enforce the policy for a
different Kubernetes network kind.

#### Use Case: Principle of least privilege policy

A cluster administrator would like disallow the use of a list of reserved labels
by default, but allow use of the labels in specific namespaces so long as the
label values are valid. 

- Define the following policy configurations:
  - policy configuration that matches all namespaces with the
    "may-use-reserved-labels" label and checks that the values are valid
  - "default" policy configuration that matches all namespaces without the
    "may-use-reserved-labels" label and disallows any reserved labels.
  - An additional policy can be used validate that only authorized roles set the
    "may-use-reserved-labels".

> See "Default configuration

Reviewer 6: given mutually exclusive labels, you don't actually need default configuration, right?

Author: Correct, if we did do default configurations, it wouldn't be unlocking functionality, it would be purely for ergonomics. Also, I haven't found any sane designs that would make this possible. I'm going to drop it.

#### Use Case: Validating native type with new field (version skew case)

Policy author wants to write a policy that validates a property of all
containers in a pod, including ephemeralContainers for versions of
Kubernetes where ephemeralContainers are available.

```yaml
  validations:
    - expression: "object.spec.containers.all(c, c.name.startsWith('xyz-'))"
    - expression: "!has(object.spec.initContainers) || object.spec.initContainers.all(c, c.name.startsWith('xyz-'))"
    - expression: "!has(object.spec.ephemeralContainers) || object.spec.ephemeralContainers.all(c, c.name.startsWith('xyz-'))"
```

> This does not work if type checking is strict. The ephemeralContainers field

Reviewer 6: `!has(object.spec.ephemeralContainers)` doesn't function properly with strict type checking?  Is there a "does this field exist in the schema" function we can use during execution?

Author: Good question.

Unfortunately no, CEL doesn't offer a way to do that. Type checking is optional and if run, it produces any errors it finds, but it does not supplement the "compiled" CEL expression with additional type information to use at runtime.

Reviewer 9: Is there any way to expose the `ephemeralContainers` field as a dynamically typed object that's exposed by default on the spec? Effectively extend the type provided with one that layers in some added metadata?

will be reported as unrecognized.

This becomes far more compact if we offer scoping:

```yaml
  validations:

>     - expression: "self.name.startsWith('xyz-')"

Reviewer 6: I don't think I find this easier to read.

Author: Added a note to the UNRESOLVED.

metadata:
  name: "policy1.example.com"
spec:
  match:
    rules:
    - apiGroups:   ["example.com"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["myCRD"]
    matchPolicy: Exact
  validations:
  - rule: "object.v1fieldname == 'xyz'"
```

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "policy1.example.com"
spec:
  match:
    rules:
    - apiGroups:   ["example.com"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["myCRD"]
      effect: Exclude
    - apiGroups:   ["example.com"]
      apiVersions: ["*"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["myCRD"]
      matchPolicy: Equivalent #is the default
  validations:
  - rule: "object.v2fieldname == 'xyz'"
```

#### Use Case: Migrating from validating webhook to validation policy

Steps:

1. Webhook is configured and in-use.
2. `ValidatingAdmissionPolicy` created with `FailPolicy: Ignore`
3. `ValidatingAdmissionPolicy` is monitored to ensure it behaves the same as te webhook (logs or audit annotations can be used)
4. `ValidatingAdmissionPolicy` is updated to `FailPolicy: Fail`

> 5. Webhook is configured with `FailPolicy: Ignore` (optional)

Reviewer 6: you would want to be sure the webhook never returns an error.  If the admission policy is equivalent, then policy would always fail before webhooks are called.

Author: That works out nicely. Added.

(issue, KEP) - requires mutation.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### CEL Integration with Kubernetes native types

While implementing [CRD Validation
Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language),
CEL was integrated with CRD structural schemas and the "unstructured" data
representation. For admission control, we also need CEL to be integrated with
the Kubernetes Go structs used to representative native API types, both for type
checking and for runtime data access.

### Writing to Status

> This enhancement proposes using status to of `ValidatingAdmissionPolicy` to

Reviewer 6: ```suggestion
This enhancement proposes using status of `ValidatingAdmissionPolicy` to
```

failure policies, type safety, advanced matching rules, ...) that will be discussed
in detail further on in this proposal.

We have divided this proposal into phases, all of which must be completed before
this feature graduates to beta. Our goal is to size the phases so that each
can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  config:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  match:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]

>       resources:   ["deployments"]

Reviewer 10: What happens if multiple resources are described here? Each config object applies to all of them?

Author: Yes, during admission, if the resource in the request is one of the resources in this array, it would be validated.

Reviewer 10: So all of these have to typecheck against the CEL expression, right?

Reviewer 3: I'm guessing the follow-up is "what about resources with different schemas?"

E.g. Someone wants to constrain workload resources so they check `CronJob`, `Deployment` and `Job`?

Reviewer 10: Yes, exactly. And that will typecheck just fine if you use common fields, which for that set might just be metadata fields. But you have to do N different typechecks to know that.

Author: The "informational type checking" covers some of the wrenches that have been thrown at our type checking plans.

Some alternatives:

- We use GVK matching in policy definitions (only) and only allow two options: (a) Specify an exact Kind or (b) allow all kinds (useful for metadata only validations). Policy bindings still use GVR matching.
- Use GVR matching but only type check if all the match resources are of the same GVK (e.g. "*/scale")
- Type check against every GVK matched by the GVR match rules.
- Don't do any type checking.

  denyReason: Invalid
  validations:
    - name: max-replicas
      expression: "object.spec.replicas <= config.maxReplicas"
      messageExpression: "'object.spec.replicas must be no greater than ' + string(config.maxReplicas)"
      # ...other rule related fields here...
```

The `spec.config` field of the `ValidatingAdmissionPolicy` specifies the kind of

> resources used to configure this policy. For this example, it is configured by

Reviewer 10: "to configure this policy" -- this is confusing, from everything else here it sounds like this is the kind which will be limited by this policy, but this sentence makes it sound like this kind is somehow input into this policy.

Reviewer 10: oh, I see, this sentence is correct and I was reading it wrong. You're putting knobs for the CEL into a separate type? Why? (I will keep reading)

apiVersion: rules.example.com/v1
kind: ReplicaLimit
metadata:
  name: "replica-limit-global.example.com"
spec:
  match:
    namespaceSelectors:
    - key: environment
      operator: Exists
  enforcement:
    max-replicas: Deny
  maxReplicas: 100
```

With this configuration, the test and global policy configurations overlap.
Resources admitted to test environment would then be checked against both policy
configurations.

To summarize, this "API Shape" separates admission policy _definition_ from
_configuration_. This has a couple advantages:

- Access to, and delegation of, policy configuration is more manageable. In
  particular, Kubernetes RBAC works well with this design.
- Without the separation, the next most obvious API shape would be to encode
  everything into a single resource, which could easily become very large and
  run into resource size limits.

##### Policy Definitions

Policy definitions are responsible for:

- Defining what validations the policy enforces and how violations are reported 
- Defining how a policy may be configured

Each `ValidatingAdmissionPolicy` resource contains a `spec.match` to declare
what resources it validates. This field is required.

> - `spec.match` constraints which resources this policy can be applied to. Policy

Reviewer 10: ```suggestion
- `spec.match` constrains which resources this policy can be applied to. Policy
```

  E.g. a CEL expression that is constrained to CREATES and UPDATES of resources
  is guaranteed the root `object` variable is never null, but a CEL expression
  that might need to evaluate a DELETE must handle the root `object` variable
  being null. See below "Match criteria" section for how match criteria is
  described.
- `spec.match` guides CEL expression type checking, see the below "Type safety"
  section for more details. 

CEL expressions have access to the contents of the `AdmissionReview` type,
organized into CEL variables as well as some other useful variables:

- 'object'
- 'oldObject'
- 'review'
  - 'requestResource' (GVR)
  - 'resource' (GVR)
  - 'name'
  - 'namespace'
  - 'operation'
  - 'userInfo'
  - 'dryRun'
  - 'options'
- 'config' - configuration data of the policy configuration being validated

See below "Reporting violations to Clients" for more detail about how the
`spec.validations` field works and how violations are reported.

##### Policy Configurations

Policy configurations are the interface between `ValidatingAdmissionPolicy`
authors and the administrators that configure the policies for clusters. The

> configurations allows policy definition authors to define how exactly a

Reviewer 10: ```suggestion
configuration allows policy definition authors to define how exactly a
```

| Generate CRDs                                     | Policy defintions contain inline OpenAPIv3 schema for config, CRD is generated with both match and config sections |

###### Alternative: Duck Typed CRDs

This is the alternative shown in the initial examples of this KEP.

Policy authors write a CRD to define how each policy is configured. E.g.:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: replicalimit.rules.example.com
  annotations:
    admission.kubernetes.io/is-policy-configuration-definition: "true"
spec:
  group: example.com
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                config:
                  type: object
                    maxReplicas:
                      type: int
  scope: Cluster
  names:
    kind: ReplicaLimit
```

The `admission.kubernetes.io/is-policy-configuration-definition` annotation
means "inject the correct .spec.match during admission and keep it up to date

> with a controller". (Suggested by deads2k). This minimizes version skew if new

Reviewer 10: Is SSA configured to treat this as atomic? CRD schema has been used as an example where you wouldn't want multiple authors in the past, lol.

Author: You're totally right, I confirmed on a local cluster that the `spec.versions` is atomic. That's inconvenient. I was naively hoping to have the managed field owners nicely separated by using a controller. 

Since SSA is no help here, is it acceptable supplement the schema via a mutating admission controller? The client that writes the schema then stays as the managed field owner, they can always safely omit the `match` part of the schema and it will be added by the mutating controller.

Reviewer 10: If you wanted to do this option, I'd make this not atomic if possible.

But I really think you should not do this option, upgrading this looks extremely hard.

Reviewer 10: If you want to do this option I would make the user include this, and barf at type-check time if your clause isn't included / isn't right.

Author: I think we have consensus to not do this option.

match criteria is added to `spec.match` and also minimizes development effort by
removing the need to manually declare the fields in CRDs.

> The main challenge with this alternative is dealing with mismatches between how

Reviewer 10: I think if we ever want to modify this match section it would be extremely difficult with this design.

Author: Atomic field problem really messes this alternative up. I've made added a bunch of notes about it and moved it below the other design.

        configurations are misconfigured. (also via a Condition?).
      - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.
      - Trigger the `FailurePolicy` on admission for resources that match the
        policy configuration.
  - If the CRD is deleted:
    - Set state to "misconfigured
    - Trigger `FailurePolicy` on all admission validations
    - Add a detailed error in the status of the `ValidatingAdmissionPolicy`.

A partial `spec.match` schema (subset of the full schema) is okay so long as
only optional fields are omitted. But any unrecognized field in the `spec.match`
would not be allowed. 

Proposed annotation: 

> Example: `admission.kubernetes.io/is-policy-configuration-definition: "true"`
> 
> Used on: CustomResourceDefinition
> 
> What a CustomResourceDefinition has the annotation set to "true", the
> OpenAPIv3 schema of all versions of this resource is modified and then
> kept-in-sync by a controller to always contain the expected schema fields of
> admission policy configuration resources.

xref: https://kubernetes.io/docs/reference/labels-annotations-taints/

Pros:

> - A single resource is used to configure both the match criteria and

Reviewer 10: This is good and bad, though? Role and roleBinding being separate is a pain until you want to use the same role in different contexts.

Author: Very true. I split out the pros and cons here.

  configuration params of a policy.

> - Provides a clear marker on every CRD that is used for policy configuration

Reviewer 10: We can and should request that people include a tag / label / annotation / etc regardless. This is not a unique pro of this option.

Author: Moved this out of pros and made a separate comment about tag / label / annotation

  that it is intended for this purpose, in a machine readable way.

Cons:

- API server must check for a wide range of error conditions and define how
  exactly it handles each of them.
- If the `spec.match` schema is incorrectly defined, CRD author might not
  realize it since they need to check the status of the corresponding

>   `ValidatingAdmissionPolicy` for any errors.

Reviewer 10: add: changing this schema in the future could be extremely difficult

Author: Added. I've included a few details about the schema fields being atomic.

> ###### Alternative: `PolicyConfiguration` kind with config reference

Reviewer 10: We should use the word "binding" for this concept somewhere, since it's analogous to roles & roleBindings. knobs & knobBindings, or policy and policyBindings?

Reviewer 10: Just to check my understanding, this design is a three object design, right? (One has the CEL expression, one has the settings/config, and one describes how settings relate to objects undergoing admission?

Author: This is a nice analog. I'll switch to "binding".

Reviewer 10: (I strongly prefer this design, in case that isn't obvious :) )

Author: Thanks for weighing in! Yes, this is the three object design.

This is the most difficult technical issue remaining. I've gone back and forth between this design and the duck type design. This design is the least magical and uses the API machinery in already well supported ways. It's more verbose (via having more objects).  The SSA problem has me leaning toward this design.

Reviewer 10: I think in practice this will often be a 2 object design, when you don't have any config that isn't in the CEL.

This is a very complex feature, I think you should heavily weight the least magical design.

I think this is the only design that lets everything be fully typechecked, with no schema manipulation?

Author: I am really happy to not have to deal with any schema manipulation!

Require two resources:

- A `PolicyConfiguration` resources contains the match criteria and a reference
  to a configuration resource.
- A configuration resource that contains only configuration (no match criteria).

Example:

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: PolicyConfiguration
metadata:
  name: "xyzlimit-scale.example.com"
  labels:
    policyConfig: xyzlimit
spec:
  match:

>     resourceRules:

Author: Yeah, this is better shown using a namespace selector. Updating.

      - apiGroups: ["*"]
        apiVersions: ["*"]
        resources: ["*/scale"]
  config:
    name: xyzlimit-scale-config.example.com
```

```yaml
apiVersion: rules.example.com/v1
kind: XyzLimit
metadata:
  name: "xyzlimit-config.example.com"
spec:
  allowed: ["a", "b", "c"]
  banned: ["x", "y", "z"]
  xyz:
    fuzzFactor: 0.8
    reticulate: true
```

Note that for policy configurations that require no parameterization, only the
`PolicyConfiguration` is needed.

The `metadata.labels.policyConfig` label on `PolicyConfiguration` is needed so
that each `ValidatingAdmissionPolicy` use a label selector to reference the

> policy configurations that configure it.

Reviewer 10: Why is this query needed? It needs to be in a field also, not just a label.

Reviewer 10: (apiserver is going to have to load all VAPs and PCs anyway so there is no need to make the query efficient)

Author: Good point. Dropping this.

Pros:

- Matching criteria is fully defined and validated in a builtin type.
- The configuration data that the policy definition is treated as an opaque
  payload from an apiserver perspective, it passes it to the policy definition
  so that CEL expressions can access it, but there is no code in the apiserver

>   that cares about the structure of the configuration CRD.

Reviewer 10: ? You can load the CRD's schema and type-check when you do the other typechecking?

Reviewer 10: (Also if this is true it should be a con, but I don't see why it would be true)

Reviewer 10: I think this alternative makes fully type checking CEL expressions the easiest.

Author: I agree about type checking. I think I failed to make a point here. I was trying to note that unlike the "duck typing" alternative, the API server is not required to have any expectations about the structure of the CRD for this design, and that makes things simpler.

I'll rephrase and add a note about type checking.

> - If later Kubernetes OpenAPIv3 supports `$ref`, there is a migration path from

Reviewer 10: Is this pro on the wrong alternative? It seems to go better in the first design alternative. I don't see how we could ever collapse this 3rd type into an embedded clause?

Author: Oops, this was misplaced. It should have been the 1st alt.

  this approach to the `$ref` approach (below).

Cons:

> - Requires two resources for each configuration of a policy

Reviewer 10: only if you need a settings type!

Author: Yes, not a real con. Dropping this.

  - Cannot atomically change the match rules and configuration of a policy configuration
- Must establish relationships between `ValidatingAdmissionPolicy` and

>   `PolicyConfiguration` resources.

Reviewer 10: This only requires a name, though. PC refers to its VAP. How is this a con?

Author: I really like the way this all falls out with "policy bindings" and separate parameter resources. I've rewritten the relevant sections to match. Hopefully others agree this is the right approach!

###### Alternative: OpenAPIv3 `$ref` in CRDs

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: replicalimit.rules.example.com
spec:
  group: example.com
  versions:
    - name: v1
      served: true
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                config:
                  type: object
                    maxReplicas:
                      type: int
                match:
                  $ref: "#/components/schemas/matchrules"
  scope: Cluster
  names:
    kind: ReplicaLimit
```

Pros:

- Match rule schema is owned by Kubernetes, so as it evolves, CRDs automatically
  pick up changes.

Cons:

- CRDs to not yet support OpenAPIv3 `$ref`s, so support would need to be added,

>   presumably this would require a separate KEP.

Reviewer 10: Yeah there is no current plan for this. A plan would need to solve the upgrade difficulty I mentioned on the first alternative and that's not easy at all, I think.

Author: Eliminating this alt.

> ###### Alternative: `/matchRules` subresource

Reviewer 10: It doesn't sound to me like this solves the problems with the first or third and instead it adds a new one?

(If we were doing this in-tree and we wanted a "no bindings" system this would be the way to go, I think. But neither are true?)

Author: Eliminating this alt too. I think everyone has agreed that it's not viable.

The idea of this alternative is to require a configuration CRD to declare that
it provides match criteria by exposing a `/matchRules` subresource:

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: replicalimit.example.com
spec:
  group: example.com
  versions:
    ... # configuration schema(s) go here
  subresources:
    matchRules:
      matchRulesPath: .spec.match
```

Pros:

- CRD explicitly opt-in to providing match criteria in a structured way.
- Follows pattern used by scale subresource to provide polymorphism across CRDs

Cons:

- One of the primary purposes of subresources is accessing/modifying a portion
  of a resource independently, which is not what we're trying to achieve.
- Kubernetes development/maintenance perspective, subresources, particularly for
  CRDs, are expensive to introduce and maintain.

> ###### Alternative: `PolicyConfiguration` kind with config embedded

Reviewer 10: No type safety for config, right? I would reject this one.

Author: I agree. Removed.

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: PolicyConfiguration
metadata:
  name: "replica-limit-prod.example.com"
spec:
  match:
    namespaceSelectors:
    - key: environment,
      operator: NotIn,
      values: ["test"]
  config:
    apiVersion: rules.example.com/v1
    kind: ReplicaLimit
    spec:
      maxReplicas: 100
```

Pros:

- Matching criteria is fully defined and validated in a builtin type

Cons:

- Embedded config needs the same treatment as custom resource, but
  reimplementing it all on an embedded resource is, at best, highly impractical
  in the apiserver as it exists today. E.g.. there is not automatic validation
  of the embedded resource.

> ###### Alternative: Generate CRDs

Reviewer 10: This alternative makes it very hard to author config, since not all the types will exist even locally before application! It's very hard to have the config management system wait for things that don't exist. There are a lot of moving pieces. I would vote for rejecting this alternative.

Author: Removed this one as well.

    kind: ReplicaLimit
    version: v1
    openAPIV3Schema:
      type: object
      properties:
        spec:
          type: object
          properties:
            maxReplicas:
              type: int
```

Pros:

- Policy author doesn't need to define the CRD. It can instead be generated for
  them.

Cons:

- Most of the cons of "Duck Typed CRDs" alternative apply, since ultimately a
  CRD is created and behaves the same.
- This implies that the policy owns the configuration. But we have use cases
  where multiple policies share the same configuration. For this model it seems
  misleading and potentially problematic to have both policies attempting to
  define the same configuration.

##### Match Criteria

During admission, the Kubernetes API server will validate the resource being
admitted against all policy configuration resources that match the resource.

While webhook match rules give a good sense of what types of capabilities might
be needed, they serve a slightly different purpose.  Webhook match rules make it
possible to avoid webhook requests, which incur latency and impact availability,
for resources that don't need to be evaluated by the webhook. CEL expressions

> have a comparatively small impact of latency and are in-process (and so do not

Reviewer 10: How many match criteria will you have to evaluate per mutating change?

Should probably set a max of some sort. That will add up eventually.

(we can compile a decision tree if we eventually need too many)

Author: I've added a section to spell out all the limits we will put in place. Ones that come to mind are:

- Max policy bindings per definition
- Max lengths for all lists in match criteria (resourceRules, namespaceSelectors, labelSelectors, ...)
- per expression CEL evaluation cost limits (we will reuse the resource limits we introduced in CRD validation)
- per policy CEL evaluation cost limits

information.  We should be careful about providing access to cross-object
information like namespace objects to webhooks since they can be stale.)

### Configurability

Consider an admission rule that disallows requests based on a blocklist.

While it is possible to inline a blocklist directly into a CEL expression as a
data literal (`!(object.metadata.name in ['blocked1', 'blocked2'])`), this quickly
becomes problematic:

- Long blocklists become unwieldy quickly in CEL expressions
- A blocklist per scope (e.g. namespace) is inconvenient express and maintain

The need to configure admission rules is common enough (see below use cases)
that we propose configuration be a 1st class concept in the API.

Since all the policy frameworks we have surveyed have configurability as a 1st
class concept, omitting it would result in either the policy frameworks not
adopting this enhancement (and sticking with webhooks) or somehow bypassing the
limitation. One possible approach would be to generate a CEL expression with the
configuration data embedded as a data literal, but we would strongly prefer not
to have policy frameworks generating a CEL expression for each possible

> configuration.

Author: Thanks, yes, this conveys the indent well. Added!

failure policies, type safety, advanced matching rules, ...) that will be discussed
in detail further on in this proposal.

We have divided this proposal into phases, all of which must be completed before
this feature graduates to beta. Our goal is to size the phases so that each
can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  paramSource:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  match:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]

>   denyReason: Invalid

Author: Yes `denyReason` assigns a status code the result of this policy if it evaluates to deny. This aligns with how admission webhooks may specify a `reason` field in deny responses.  I've clarified in the KEP near this example as well.

  validations:
    - name: max-replicas
      expression: "object.spec.replicas <= params.maxReplicas"

>       message: "object.spec.replicas must be no greater than {1}"

Author: This was my first choice (see the alternatives section at the bottom).

I could still be convinced to go another way here. The current down sides of using CEL directly seem to be:

- This example requires explicit cases, e.g. `"object.spec.replicas must be no greater than " + string(params.maxReplicas)`
- a plain string message had to be double quoted, e.g. `"'this is a simple message'`. Note that for CRD validation we only allow messages to be plain strings, so this creates an inconsistency. CRD validation would allow `message: "this is a simple message"` but admission would not.

The second one is more concerning for me than the first. We could offer both a `message` and `messageExpression` field (mutually exclusive)..

>       message: "object.spec.replicas must be no greater than {1}"

Reviewer 9: That's right, I was thinking of a feature I'd used to treat fields as string literals or CEL. If parsed as valid CEL, I treated it as such, then compiled. Otherwise, I converted the string to a valid CEL string literal and turned that into the program. It's a bit of a hack, and I'm not sure it would work in this case, but it certainly makes the field easy to use. 

I'm also totally open to exposing a `string.format(string, list(dyn))` or `string.format(string, map(string, dyn))` function. I'd probably try to use `fmt.Sprintf` style strings since the format is well-implemented and relatively consistent across C++, Java, and Go. If it can convert to a string, then the `string(<value>)` will be applied with special handling for `list` and `map` values.

>       message: "object.spec.replicas must be no greater than {1}"

Author: SGTM, I'll switch over to `message` and `messageExpression` and make a note about `string.format()` (could be receiver fn? `"%d must be less than %d".format([value, max])`)

      messageArgs: ["params.maxReplicas"]
      # ...other rule related fields here...
```

The `spec.paramSource` field of the `ValidatingAdmissionPolicy` specifies the
kind of resources used to parameterize this policy. For this example, it is
configured by `ReplicaLimit` custom resources. Note in this example how the CEL
expression references to the parameters via the CEL `params` variable, e.g.
`params.maxReplicas`.

> `spec.match` specifies what resources this policy is designed to validate. This

Reviewer 9: Is this more like a selector described at the policy level rather than the instance? If so, you might be able to omit it by focusing such extensibility on the policy binding instead.

Reviewer 10: Or give the two match clauses different names so it's clearer what is going on and how they interact. E.g. "matchResources" and "matchObjects"

Author: @TristonianJones this is a selector at the policy level. The primary goal here is to allow the policy definition author declare the conditions where their policy makes sense. If the policy author is writing a policy to validate the contents of Pod resources, they benefit by specifying that since they can then write CEL expressions that can safely assume that the object is the type they expect and that the object is not null (which can be the case if they don't establish with these match rules that the operation must be "CREATE" or "UPDATE" but not "DELETE").
 

@lavalamp I like the idea of distinguishing them, they are semantically different.  How about `matchConstraints` for policy definitions and `matchResources` for policy bindings?

Author: I've applied this. We can finalize exact field names in API review.

also guides type-checking, see the "Informational type checking" section for
details.

The `spec.validations` fields contain CEL expressions. If an expression
evaluates to false, the validation check is enforced (the below `PolicyBinding`
configures the enforcement).

This is a "Bring Your Own CRD" design. The admission policy definition author is
responsible for providing the `ReplicaLimit` parameter CRD.

To configure an admission policy for use in a cluster, a binding and parameter
resource are created. For example:

```yaml
# Policy binding
apiVersion: admissionregistration.k8s.io/v1
kind: PolicyBinding
metadata:
  name: "xyzlimit-scale.example.com"
spec:
  policy: "replicalimit-policy.example.com"
  params: "replica-limit-test.example.com"
  match:
    namespaceSelectors:
    - key: environment,
      operator: In,
      values: ["test"]

>   enforcement: [Deny]

Author: This is a good question. The API feels a bit clunky to me here, but I'll try to explain and maybe we can smooth it out.

_If_ a policy evaluation for a resource results in `Deny` _then_ the `denyReason` is used to populate the status code of the (rejected) admission request.

Policy bindings get to decide enforcement, but the policy definition gets to decide the denyReason.

  null. See below "Match criteria" section for how match criteria is described.
- `spec.match` guides CEL expression type checking, see the below "Type safety"
  section for more details. 

CEL expressions have access to the contents of the `AdmissionReview` type,
organized into CEL variables as well as some other useful variables:

- 'object'
- 'oldObject'
- 'review'
  - 'requestResource' (GVR)
  - 'resource' (GVR)
  - 'name'
  - 'namespace'
  - 'operation'
  - 'userInfo'
  - 'dryRun'
  - 'options'
- 'config' - configuration data of the policy configuration being validated

See below "Reporting violations to Clients" for more detail about how the
`spec.validations` field works and how violations are reported.

##### Policy Configuration

`PolicyBinding` resources and parameter CRDs together define how cluster
administrators configure policies for clusters.

Each `PolicyBinding` contains:

- `spec.policy` - A reference to the policy being configured
- `spec.match` - Match criteria for which resources the policy should validate
- `spec.params` - Reference to the custom resource containing the params to use
  when validating resources 

> - `spec.enforcement` - How validation failures should be enforced

Reviewer 9: I think enforcement should have a narrower scope as it looks like it's combining many different policy types and concepts in one here. 

Consider that the admission policy effect is always to reject the request, but the error code determines whether the reason for the rejection is user-visible or not. Should this decision about the error code belong with the person who configure the policy or the person who configures the binding? My bias is toward the person who configures the policy. 

Now, let's say the person creating the policy binding wants to control rollout of the binding, shifting it from a dry-run state which audit logs rejections or possibly audit logs differences between the itself and another binding. That's a very reasonable decision to put into the hands of the person who creates the binding. However, I wouldn't call this `enforcement`, but instead `rollout`. If left unspecified the `rollout` would be `live`.

Ideally, the decision to reject the request is separate from the decision to deny (or grant) the request. Until such. a policy exists, I would consider always auditing permission denied responses (as opposed to invalid argument responses).

Reviewer 10: I agree that either params or binding is likely a better place for this. But we have to define what happens when multiple params are bound, or one param applies via multiple bindings.

Reviewer 3: It might be worth reconsidering a more complex syntax for `enforcement` to allow a bit more decoupling between "signal sent by policy" and "action taken by enforcement point"

Example:

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: PolicyBinding
metadata:

>   name: "xyzlimit-scale.example.com"

Reviewer 10: This name can't work since there can be multiple.

Author: Oops, fixing the examples.

spec:
  policy: xyzlimit-scale.example.com

>   params: xyzlimit-scale-settings.example.com

Reviewer 10: If we're only stating the resource, is this necessary at all? Can't we get that from the policy?

Author: I was expecting a 1:1 relationship between a policy binding and a param CR. (and a 1:1 between policy binding and policy definition).

Reviewer 10: I expect multiple bindings per param CR. If no other reason, consider that there will eventually be a namespaced binding if I understood phase 2 correctly.

For RBAC, without looking, I'd bet there's maybe 5x the number of bindings as roles in a typical cluster, I expect this to be similar, maybe not quite as extreme because policy & params will serve some of that need.

Reviewer 10: I think a diagram with arrows or at least a text explanation of the relationship between these objects would be helpful.

Author: Oh yes, I'm sorry, I was tired when I wrote the above.

relationships should be:

- 1:N policy definition to policy binding
- 1:N param CR to policy binding

?

I was thinking of putting the references from each policy binding to one policy definition and one param CR to achieve this.

Author: Oh, @lavalamp were you thinking it should be {1:N definition:binding} and all params are applied to all bindings?

Reviewer 10: I was last night, but now I think that was wrong -- I hadn't realized that a single params object can work with multiple policy objects. So you need to state both, I think.

  match:
    namespaceSelectors:
    - key: environment,
      operator: Exists
  enforcement: [Deny, Audit]
```

Each parameter CRD defines the custom resources that are referenced by the
`spec.params` field of `PolicyBinding` resources. Example:

```yaml
apiVersion: rules.example.com/v1
kind: XyzLimit
metadata:
  name: "xyzlimit-settings.example.com"
allowed: ["a", "b", "c"]
banned: ["x", "y", "z"]
xyz:
  fuzzFactor: 0.8
  reticulate: true
```

Note that for policies that require no parameterization, only the
`PolicyBinding` is needed.

See "Alternatives considered" section for rejected alternatives. This design was
selected because:

- Matching criteria is fully defined and validated in a builtin type.
- Type checking is straight forward.

> - Policy parameterization is separated from the policy binding, allowing for

Reviewer 9: I believe the above two points are also true if you have a (policy, binding) tuple where the binding contains the parameters inlined. Is there a benefit to being able to version the parameters type definition over time? I feel like perhaps there are other benefits which just haven't been enumerated.

Reviewer 10: Combining binding and parameters gives you an object with two different owners for different parts of the schema, which mostly doesn't work (makes type safety very hard). All of the alternatives tried to do that in some way.

Author: The 'alternatives considered' has a bunch of binding + parameters inlined designs that didn't pan out. I've really warmed up to this (policy, binding, params) split over time. It makes some rollout workflows much easier (e.g. adding a new validation expression to a policy).

Reviewer 9: @jpbetz I'm mostly looking for the benefits to be spelled out more clearly. If the schema ownership is a good point to separate concerns, then that's probably worth mentioning.

Author: SG, I've added this to the list as well as some other advantages that have been mentioned.

  well abstracted parameterization types to be used by applied in different ways
  by multiple policies.

##### Match Criteria

During admission, the Kubernetes API server will validate the resource being
admitted against all policy configuration resources that match the resource.

While webhook match rules give a good sense of what types of capabilities might
be needed, they serve a slightly different purpose.  Webhook match rules make it
possible to avoid webhook requests, which incur latency and impact availability,
for resources that don't need to be evaluated by the webhook. CEL expressions
have a comparatively small impact of latency and are in-process (and so do not
have the same impact to availability).

For CEL expressions, the primary benefits of match criteria are:

- Match criteria establishes bounds on what sort of admission requests the CEL
  expressions must consider. The CEL expressions can be written knowing that the

>   match criteria has filtered out requests that is does not need to consider.

Reviewer 9: Perhaps, this is an argument in favor of using `match` to simplify CEL expression authoring? Could you leave `match` empty and replicate the arguments inside the CEL expression to achieve the same result? Would having the ability to declare and use local variables also achieve a similar objective when it comes to simplicity of expressions? You run the risk of having two ways of doing the same thing if `match` becomes sufficiently complex. Or does `match` assist with policy indexing and as a side-effect simplify the CEL expression? I think if `match` exists it needs to be very simple and needs to help achieve something that might not be easy to do with CEL directly.

Reviewer 10: Match will be understandable to existing kubernetes authors / configurers. It's going to be necessary to run the match criteria extremely often and it needs to be done in go code. My comment has been lost, but if necessary we need to be able to compile a decision tree about what policies need to be checked given metadata.

Author: We did consider not having any "YAML matching" for this feature and instead pushing all matching into CEL. The main deciders for me were:

- Kubernetes already has resource matching as a well established concept
- Match criteria can be built indexed/accelerated/built into decision trees
- Match criteria can evaluate only to true/false. There is no 'error' case to consider
- Match criteria can be used to guide static typing. If the match is for v1.Deployment, we know ahead of runtime what type the `object` variable is

Reviewer 9: @lavalamp @jpbetz would you mind including these key pieces of information in the text above? These sound like much stronger arguments than being familiar and helpful to CEL since they also affect performance and safety in ways that cannot easily be achieved with pure CEL.

Reviewer 10: +1 this is probably worth updating the text over.

| operation (HTTP verb)                    | WH/Audit/P&F                 | phase 1                                |
| exclude                                  | Audit (level=None)           | phase 1                                |
| apiVersion + kind                        |                              | phase 1                                |
| NonResourceURLs                          | Audit/RBAC/P&F               | No                                     |
| user/userGroup                           | Audit                        | phase 2 see "Secondary Authz" section  |
| user.Extra                               | (in WH AdmissionReview)      | phase 2 see "Secondary Authz" section  |
| permissions (RBAC verb)                  | RBAC                         | phase 2 see "Secondary Authz" section  |

WH = Admission webhooks, P&F = Priority and Fairness

Match criteria must be declared in the `spec.match` field of `PolicyBinding`
resources (see `ReplicaLimit` in the above example) and will be declared with
API types in a format similar to admission webhooks, P&F, RBAC and Audit, but
with improved support for exclude matching.

*Excluding*:

> Exclude support make it possible to do things like validate all namespaces

Reviewer 9: nit: s/make/makes

> Exclude support make it possible to do things like validate all namespaces

Reviewer 10: ```suggestion
Exclude support makes it possible to do things like validate all namespaces
```

> except `kube-system`. This is difficult to support with out direct exclude

Reviewer 10: ```suggestion
except `kube-system`. This is difficult to support without direct exclude
```

    apiVersions: ["*"]
    operations:  ["*"]
    resources:   ["*"]
```

This forces the webhook configuration author to explicitly declare what
they intend to match.

The same principle applies here but with one major difference-- if a policy
definition has a match rules for apiGroup + resource + operation, then all
bindings of that policy are already constrainted to that apiGroup + resource +
operation match.

Take, for example:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
...
spec:
  paramSource: ...
  match:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
```

Since this policy is constrainted to create/update of deployments. Policy
bindings don't need to repeat this constraint. This would be sufficient:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: PolicyBinding
...
spec:
  match:
    namespaceSelectors:
    - key: environment
      operator: Exists
 ...
```

We can enforce this by requiring:

- Policy definitions match rules are validated to match one or more apiGroup +
  resource + operation.

> - Policy bindings are not required to match apiGroup + resource +

Reviewer 3: But they may? In cases where a policy definition matches more than one group/resource/operation and the binding author wants to further scope that down?

Reviewer 10: +1 are they allowed to do this? I can't think of a good reason so an example would help.

Reviewer 10: (now that I see @TristonianJones' comment) The matches are doing different things; the policy needs a match that defines resources, so that the CEL expressions can be evaluated for type safety. The binding doesn't need to (and IMO shouldn't be able to) repeat that. But it needs to decide based on metadata (name, namespace, labels, etc) whether a policy needs to be run on a given object.

Author: > are they allowed to do this?

For group/resource/operation, the way the KEP is written the binding could further restrict it down. There are practical applications for being able to do this.  E.g. for a rollout of a policy, being able pick exactly what resources will transition from warning to deny for a policy can be achieved if the policy bindings (which pick enforcement) can also pick which resources that enforcement should apply to.

+1 to @lavalamp's comment about type safety.

supports wildcards could be used.

Match Policy:

`MatchPolicy` will work the same as for admission webhooks. It will default to
`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
definitions for different versions of CRD" for an explanation of why we need
`MatchPolicy`.

xref:

- https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
- https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/
- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#resourcepolicyrule-v1beta1-flowcontrol-apiserver-k8s-io
- https://github.com/open-policy-agent/gatekeeper/blob/a1add93b0beb5c48eb92a6a2eb5ee7d21551a1b6/pkg/mutation/match/match.go#L22-L56

#### Reporting violations to Clients

This section focuses on how information is reported back to clients in
when validations fail.

Goals:

- Feature parity with `AdmissionReview`
  - Support allow/deny result, warnings, audit annotations
  - Support for reasons/codes
- Ability to format message strings

High level proposal:

- Each policy may set a `denyReason` and/or `denyCode`
- Each validation may define a message:

>   - `message` - may contain `{1}` format args that are supplied by

Reviewer 9: Any reason why the message can't be a CEL expression itself?

Author: We looked at this as an option, see "alternatives considered" and my above, I'm happy to reconsider this decision as I don't feel super strong about the alternatives. Ideally we come up with a plan that works for CRD validation as well.

Author: I've added a UNRESOLVED to this section so we don't approve this design until we decide on an approach here.

Author: Switched to `message` and `messageExpression`

>     `meesageArgs: [<cel expression>, <cel expressiion>, ...]`

Reviewer 10: ```suggestion
    `meesageArgs: [<cel expression>, <cel expression>, ...]`
```

Reviewer 10: ```suggestion
    `messageArgs: [<cel expression>, <cel expressiion>, ...]`
```

below has this typo also

    - If `meesage` is absent, `expression` and `name` will be included in the

>       failure message

Reviewer 9: nit: s/meesage/message

    - If any of the arg CEL expressions fail: `expression` and `name` will be
      included in the failure message plus the arg evaluation failure
- allow/deny, warnings, audit annotations will be collectively referred to as
  `enforcement`, which will have the following levels:
  - `Deny`,
  - `Warn`

>   - `Audit`

Reviewer 10: I was hoping for some sort of canary-ing here, like so you know if it will destroy your cluster before you turn it on.

Author: An orchestration layer above must create multiple policy bindings, one can be the canary.  I'm not convinced the matching criteria is sufficient for fine grained canary-ing though.

Also, we will have metrics. I suppose `enforcement: []` would need to result in metrics only for a binding, but that would be far from obvious to users.

Reviewer 10: I was expecting canary-ing by modifying the binding object to make it more strict, which implies the need for a observe-only mode or something like that.

OTOH you could argue that the CEL policy is the thing that needs canary-ing and that kind of thing should go in the policy object.

But I think there must be some way to roll out a policy to less than 100% severity on 100% of the cluster.

Author: @TristonianJones and I discussed this offline. I've added a UNRESOLVED to move the responsibilities around and make the binding author responsible only for selecting between things like "Dry-Run" or "Canary" and "Enabled", and have the policy definition author responsible for what enabled means (Deny, audit various things, ...)

Reviewer 3: per an above comment, I think if we can convince ourselves that the policy definition author has reason to want some signal as to the type of action that should be taken, we should figure out a more nuanced protocol that decouples the two parties, since the binding author *definitely* has interest in enforcement action, outside of rollout.

Take, for example, setting a policy to `Warn` not because it's a new policy, but b/c this is a dev cluster and you want to notify but not obstruct development.

Another example: forcing audit-only enforcement for critical namespaces to prioritize availability while preserving auditability.

Both of these situations are ones where a cluster admin (who I think is more likely to be writing bindings) has the necessary context, and a policy definition author (who may not even be at the same company as the cluster admin) wouldn't.

| CRD is deleted                                         | Nothing to type check against, but also means there are no coresponding custom resources |
| CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
| Validation of an aggregated API server type            | Main API server does not have type definitions |

Due to these complications, we have decided to evalute CEL expressions
dynamically. Informational type checking will be provided (except for aggregated
API server types), but will be surfaced only as warnings. See "Alternatives
Considered" section for details of all the alternatives we reviewed when
selecting this approach.

Type checking is still performed for all expressions where a GVK can be matched
to type check against, resulting in warnings, e.g.:

```yaml
...
status:
  expressionWarnings:
    - expression: "object.foo"
      warning: "no such field 'foo'"
```

#### Failure Policy

For in-process validation there is no remote request, so errors should be

> deterministic. We believe this significantly reduces the risk of "fail closed"

Author: This is true. I'll drop this paragraph. I think this reads sanely without it.

admission control as compared to webhooks.

Because failure policy is most often selected based on the need to guarantee
enforcement, we will default failure policy to "fail" and allow it to
be configured on a per-policy basis:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
spec:
  ...
  failurePolicy: Ignore # The default is "Fail"
  validations:
    - expression: "object.spec.xyz == params.x"  
```

#### Safety measures

To prevent clusters from being put into a unustable state that cannot be
recoverd from via the API, admission webhooks are not allowed to match
`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.

We will extend this approach:

> - `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`

Reviewer 10: I think you should also outlaw matching param objects or bindings!

Author: Good catch! Fixed.

- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
  `MutatingWebhookConfiguration`.

Note that this does allow `ValidatingAdmissionPolicy` to match
`ValidatingWebhookConfiguration`.

Note: In the future we may further loosen this up and allow admission
configuration to intercept/guard writes to admission configuration while
preventing deadlock - Add feature to configure a set of webhooks to intercept
other webhooks https://github.com/kubernetes/kubernetes/issues/101794.

Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
higher level. This could be added later.

#### Default Validations

A `defaultValidations` field will allow policy definition authors to define
expressions which are evaluated when the policy definition matches a resource

> but no policy configurations match that resource.

Reviewer 10: what happens if a multiple param objects are bound, or one param object is bound several times (via separate bindings that match in different ways)?

Author: Ooh, I didn't rethink this after splitting binding and config. We don't need to evaluate the CEL expressions multiple times for the same params since the result will always be the same. But the bindings may have different enforcement settings, so we will need to apply those correctly and attribute all violations to the appropriate binding. I'll add an implementation note.

Behavior:

- `defaultValidations` do not have access to params.
- `spec.paramSource` is optional only when `defaultValidations` are present
- (`spec.paramSource` and `spec.validations` must both be either present or absent)
- When a policy has `defaultValidations`, it must also declare the enforcement
  level of those validations

> A "singleton" (aka standalone) policy can be defined as:

Reviewer 10: Probably need a "default binding" section to make this truly standalone now.

Author: Yes, this it out-of-date with our most recent changes. Fixing.

All these capabilities are required before Beta, but will not be implemented
in the first alpha release of this enhancement due to the size and complexity
of this enhancement.

#### Namespace scoped policy binding

For phase 1, policy bindings were only allowed to be cluster scoped. We can
support namespace scoped policy bindings as follows:

- Add a `NamespacePolicyBinding` resource.
- If the parameter resource is namespace scoped, it implicitly matches
  resources only in the namespace it is in, but may further constrain what
  resources it matches with additional match criteria.

Benefits: Allows policy of a namespace to be controlled from within the
namespace. As an example, ResourceQuota works this way.

Details to consider:

- Should a policy support both cluster scoped and namespace scoped binding? If
  so how? It would need two different parameter CRDs (since a CRD must either be
  cluster scoped or namespace scoped, not both).

#### CEL Expression Composition

##### Variables

<<[UNRESOLVED jpbetz, ?? ]>>
cel-policy-template "terms" are very convenient. But we don't have consensus on
if they are sufficiently essential. Need to either agree on something or drop
this.

> <<[/UNRESOLVED]>>

Reviewer 3: Benefits of variables:

* Simpliify computation by evaluating complex expressions once and storing the result in a variable for reuse
* Avoid the need to duplicate brittle and repetitive logic, such as field existence checks
* May obviate the need for the "scope" section below
* Avoids copy/pasting subexpressions in CEL expressions (I don't believe CEL lets you define variables?)

There might be more, but the above seem pretty persuasive to me.

Reviewer 9: I think the readability arguments are sufficient to warrant the introduction of terms. There is some added complexity carried with doing so in that these features aren't part of the standard CEL-Go library, but can be orchestrated outside it.

Reviewer 9: I think that `scopes` should be omitted in favor of bind variables as, they're nice, but basically a way of removing the `all`, `exists` macro from the expression. In other words the scope replicates existing CEL behavior, but the term does not.

Author: I think I'm sold on this. I've dropped scopes.

Any strong preference for what we call them? `variables`, `terms`, something else?

Each CEL "program" is a single expression. There is no support for vaiable
assignment. This can result in redundant code to traverse maps/arrays or
dereference particular fields.

We can support this in much the same way as cel-policy-template if we offer an
ordered list or variables. These can be lazily evaluated while the validation
expressions are evaluated (cel-policy-template does this). The results can also
be memoized to avoid repeated evaluations if they are shared across validations.

```yaml
  variables:
    - name: metadataList
      expression: "spec.list.map(x, x.metadata)"
    - name: itemMetadataNames
      expression: "metadataList.map(m, m.name)"
  validations:
    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
    - expression: "itemMetadataNames.exists(name, name == 'required')"
```

##### Scopes

<<[UNRESOLVED jpbetz, ?? ]>> 

> This is not yet agreed on, and some reviewers have found it more

Reviewer 3: IMO I would include "bind variables" and omit "scopes", since they share functionality but scopes is less flexible than bind variables and adds a second path traversal language.

Having some kind of forEach for binding variables might be necessary to simplify expressions that traverse nested collections (e.g. all ports in all containers).

Reviewer 10: I at least agree you should pick one way to do this, not two or three.

Author: I'll focus on "variables" and drop scopes. This is also phase 2 so we have an additional cycle to refine the proposal.

difficult to understand validations using this feature than validations that do
not.

> Also, the "bind varables" idea is maybe too ambitious?

Reviewer 3: What makes "bind variables" ambitious? CEL Policy templates already has this behavior.

Reviewer 3: nvm, I see this is binding variables WRT scopes, not variables as defined in the "Variables" section immediately above.

Reviewer 9: I would omit scopes. There has been an ask for them in the past, but usually this signals the need for a separate policy.

Author: I've dropped scopes out. I've kept in content in the "alternatives considered" section for reference.

attached to any schema in the OpenAPIv3. The validation rules are scoped to
whatever location in the OpenAPIv3 they are attached.

We propose using a simple schema path format. The purpose of the path is to
uniquely identify a schema from the root of a CRDs OpenAPIv3 schema.

Example:

```
spec.initContainers         # Schema of initContainers array
spec.initContainers[*]      # Schema of the items of the initContainers array
spec.initContainers[*].name # Schema of the name of initContainers
```

For example, to validate all containers:

```yaml
  validations:
    - scope: "spec.containers[*]"
      expression: "scope.name.startsWith('xyz-')"
      message: "{1} does not start with 'xyz'"
      messageArgs: ["scope.name"]
```

To make it possible to access the path information in the scope, we can offer a
way to bind varables to the map and list indices in the path, e.g.:

```
spec.x[xKey].y[yIndex].field
```

```yaml
  validations:
    - scope: "x[xKey].y[yIndex].field"
      expression: "scope.startsWith('xyz-')"
      message: "{1}, {2}: some problem"
      messageArgs: ["scopePath.xKey", "scopePath.yIndex"]
```

Prior art:

- cel-policy-template's offer a

>   [`range`](https://github.com/google/cel-policy-templates-go/blob/master/test/testdata/map_ranges/template.yaml)

Reviewer 3: FWIW `range` actually binds variables to the value of a field at the range, the CEL expression isn't scoped.

This is significantly more powerful, as it allows you to perform checks like "only containers named 'server' have a port named 'web' that listens on port '80'"

To do the same with scope expressions,  you'd need fancier syntax (e.g. `spec.initContainers[!~^server$]` to check a regex or similar), probably something best left to CEL.

Reviewer 9: I like `range`, but think it's good as a post-alpha feature only because it's possible that some of what's offered in `range` may become a part of core CEL. 

One deficit of CEL is that macros only give you access to the list element value, but not the index (reverse for maps), and if the key, value pair were available in both contexts, then you can express what `range` does inside CEL. Currently, you can't.

Author: I think variables are good enough in the short term. I'll make note of range in the future work section.

<<[/UNRESOLVED]>>

- Configurable admission blocking write requests made internally in
  kube-apiserver during server startup (like RBAC default policy reconciliation)
  making it impossible for a server to start up healthy. (This is not specific to CEL?)
- Ability to skip specific resource types - Admission Controller Webhook
  configuration rule cannot exclude specific resources:
  https://github.com/kubernetes/kubernetes/issues/92157

#### Aggregated API servers

Main complications (provided by @liggitt):

- The API server validating/persisting the ValidatingAdmissionPolicy
  instances isn't the same one serving aggregated types, so wouldn't necessarily
  have schema info to check type safety.

- The aggregated API server is responsible for enforcing admission on its custom
  types, so the implementation that reads ValidatingAdmissionPolicy
  instances and enforces them would have to live in k8s.io/apiserver and be
  active in aggregated API servers to enforce admission on aggregated types
  effectively (same as admission webhooks today).

Plan:

- Do not offer type checking for aggregated types.
- Support ValidatingAdmissionPolicy in aggregated API servers.

> #### Short Circuiting

Reviewer 10: I think you should wait for a clear example use case before considering this.

Author: SGTM. I'll shelve this one.

Matching defined in YAML is not as expressive as CEL functions. Any additional
filtering of objects that cannot be expressed in the match rules will need to be

> handled as "second pass" match filtering in CEL.

Reviewer 9: You could make the argument that a match expression should be handled as CEL and there could be special functions exposed to handle common match patterns. I think that would be the simplest tradeoff.

Author: This section got dropped. xref https://github.com/kubernetes/enhancements/pull/3492#discussion_r981643590 for CEL for match expressions.

Latency metrics should be per {policy, validation expression}. The next level of
granularity would be {policy, binding, validation expression}, but that
the number of biindings can become quite large, so let's limit it to
{policy, validation expression} for now.

- xref: [Metrics Provided by OPA Gatekeeper](https://open-policy-agent.github.io/gatekeeper/website/docs/metrics/)
- xref: [Admission Webhook Metrics](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#admission-webhook-metrics)

### User Stories

In addition to "User Stores", see below "Potential Applications" for a list of
known applications and their use case requirements.

#### Use Case: Singleton Policy

User wishes to define a simple policy that required no parameters. They don't
want to create parameter CRD since what they're doing can be expressed quite
simply in a single CEL expression.

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  singletonPolicy: true
  match:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]

>       resources:   ["deployments"]

Reviewer 9: Consider that the equivalent CEL match expression would probably be 

```go
apiGroups.hasAny(["apps"]) 
&& apiVersions.hasOnly(["v1"]) 
&& operations.hasAny(["CREATE", "UPDATE"]) 
&& resources.hasOnly(["deployments"])
```

You could also treat this as a special function:

```
matchResources({
   "apiGroups": ["apps"],
   "apiVersions": ["v1"]
   "operations":  ["CREATE", "UPDATE"]
   "resources":   ["deployments"]
})
```

Reviewer 9: A recurring theme in this KEP is that there are needs to match or select against data available to K8s which might not be easily expressed as pure logic in CEL. You could create something separate, or you could expose the functionality to CEL. If you expose it to CEL, it's easier to combine with other predicates and will hopefully simplify the number of fields and their possible interactions.

Author: Regardless of outcome of https://github.com/kubernetes/enhancements/pull/3492#discussion_r981643590 there might need to be functions in CEL to process matching logic.  For example, there are many uses of the `labelSelector` match object throughout k8s, and having CEL be able to process them could be very useful. I've added a section in phase 2 about CEL function library and add this there.

Reviewer 9: My concern is that you might end up needing a `match` block as it's understood today and a secondary `matchExpresion` to adequately express alternatives among the validation expressions since sometimes what you really want is 'no decision' rather than a trivially true one; however, it can probably wait.

Author: This may happen. A lot of webhooks do more fine grained matching in the webhook.

Given that webhooks already have "yaml matching", even if we do improve matching to use more CEL in the future, I don't know if it does much harm to have CEL admission also using "yaml matching". It might be better to keep both as consistent as possible and extend both to support CEL matching if we go that way.

I'm okay with deferring this. I've added a note in the future work section.

  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
  and oldObject in AdmissionRequest)
- Configurability, as motivated above.

There are also lots of additional capabilities (response message formatting,
failure policies, type safety, advanced matching rules, ...) that will be discussed
in detail further on in this proposal.

We have divided this proposal into phases, all of which must be completed before
this feature graduates to beta. Our goal is to size the phases so that each
can be completed in a single Kubernetes release cycle.

### Phase 1

#### API Shape

Before getting into all the individual fields and capabilities, let's look at the
general "shape" of the API.

This enhancement introduces a new `ValidatingAdmissionPolicy` kind.

Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
The resource contains the CEL expressions to validate the admission policy and
declares how the admission policy may be configured for use.

For example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1

> kind: ValidatingAdmissionPolicy

Reviewer 11: Since we require ValidatingAdmissionPolicy, PolicyBinding, parameter CRDs and param configuration to be in place for a complex validation, is there a specify order we required(besides CRD has to be in place before a CR created)? Like Policy configuration has to be in place before the policy definition? I heard "we don't want to force a specific order" and not quite convinced. Also curious When to process the validation and check if all the dependencies in place if we don't ask for order?

Reviewer 11: Thanks for the offline sync @jpbetz . 
Considering the kubectl apply behavior, not forcing a specific order make sense and we have to handle the violation properly. The scenarios would be:
- PolicyDefinition is in place
  - PolicyBinding before param: Needs to be handled properly
  - Policy param before binding: ok
- PolicyDefinition is not in place, it does not hurt in either order 
I guess it is ok to leave bunch of PolicyDefinition there without validation process running.
Since new API is introduced with the feature, it might worth to think about feature support, conversions, backward capability, upgrade/rollout carefully before going to beta.

Author: The main ordering problem is when a binding is created before the param CR it requires. If we did check for existence of the param CR and reject the binding write if it is missing, then we would solve the ordering problems but also impose ordering between resources and have to deal with the consequences to `kubectl apply -f <directory-full-of-yaml>`.

I'd prefer to keep with the mitigations mentioned in the KEP and not require ordering, but instead handle broken binding objects in a consistent way.

Good news is that higher level orchestration (e.g. policy engines) can make sure to order things in safe ways so long as we give them the right primitives.

Reviewer 10: Yeah in general it's better to let e.g. controllers retry a few times than to force users to create objects exactly in the right order. In this case it does seem like the admin should get some feedback that the policy is actually being enforced instead of disabled due to configuration error. I think that's something to add to the kep another day if the Audit / Warn settings aren't sufficient.

  assumptions. E.g. a CEL expression that is constrained to CREATES and UPDATES
  of resources is guaranteed the root `object` variable is never null, but a CEL
  expression that might need to evaluate a DELETE must handle the root `object`
  variable being null. See below "Match criteria" section for how match criteria
  is described.
- `spec.matchConstraints` guides CEL expression type checking, see the below
  "Type safety" section for more details. 

CEL expressions have access to the contents of the `AdmissionReview` type,
organized into CEL variables as well as some other useful variables:

- 'object'
- 'oldObject'
- 'review'
  - 'requestResource' (GVR)
  - 'resource' (GVR)
  - 'name'
  - 'namespace'
  - 'operation'
  - 'userInfo'
  - 'dryRun'
  - 'options'
- 'config' - configuration data of the policy configuration being validated

See below "Reporting violations to Clients" for more detail about how the
`spec.validations` field works and how violations are reported.

##### Policy Configuration

`PolicyBinding` resources and parameter CRDs together define how cluster
administrators configure policies for clusters.

Each `PolicyBinding` contains:

> - `spec.policy` - A reference to the policy being configured

Author: Unless I'm misunderstanding, this can be supported via:

- policy A <- binding X -> params 1
- policy B <- binding Y -> params 1

> - `spec.policy` - A reference to the policy being configured

Reviewer 3: This is true, my point is that it would require multiple bindings. I.e. adding a policy def that extends an existing policy doesn't get it enforced for "free".

If needed, maybe that could be added later by not specifying `spec.policy` or similar, but worth highlighting the tradeoff.

> - `spec.policy` - A reference to the policy being configured

Reviewer 10: I think the params object can't reference the policy object, since it could work with multiple.

The policy object has to reference the params *type* (but not a specific param instance) in order to type-check.

So yeah, it seems like you have to state both in order to have a well defined binding.

Default behavior of e.g. "blank means all param objects" is not that user friendly, if we absolutely need it I would make it require a "*" or something.

> - `spec.policy` - A reference to the policy being configured

Reviewer 3: Although... without a params object, you also can't have multiple policy objects reference the same config (since this is done via listening to a common kind).

Does this imply creating empty params objects for supporting this functionality?

Or does it mean that bindings must double-cite and we lose the "multiple policies read the same config just by existing" behavior?

Or does bindings-cite-policies become optional, which makes for a bit more heterogeneity in terms of the possible config footprints?

    `messageArgs: [<cel expression>, <cel expression>, ...]`
    - If `message` is absent, `expression` and `name` will be included in the
      failure message
    - If any of the arg CEL expressions fail: `expression` and `name` will be
      included in the failure message plus the arg evaluation failure
- allow/deny, warnings, audit annotations will be collectively referred to as
  `enforcement`, which will have the following options:
  - `Deny`,
  - `Warn`
  - `Audit`
  - (metrics will be emitted regardless of which enforcement options are set, so
    `enforcement: []` may be useful)
- Policy definitions identify validation expressions by name so that if the
  enforcement is `Audit` the name can be used as the audit annotation key.
- Policy bindings set enforcement options

Example policy definition:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  ...
  denyReason: Invalid
  vars: # MUST BE A REAL LIST WITH ORDERING
    - containers: "self.containers + self.initContainers"
    - foo: "container.size()"
  validations:

>     - expression: "self.name.startsWith('xyz-')"

Reviewer 9: Do you short-circuit the validation expressions blocks or aggregate the violations messages together? My guess is that you accumulate the validation violations, but it's probably worth spelling out and whether `denyReason` affects the evaluation behavior. For example a `denyReason: Deny` could potentially short-circuit (and this also depends on whether audit logging is always required on deny).

Reviewer 10: This is a good question, deny/admit and short-circuit/continue seem like orthogonal options and maybe there should be two fields. Maybe more. API types in the KEP are never final, that happens when we do a detailed api review. In fact maybe we should just expect to change this after we do an alpha and see what some real-ish policies actually want to do.

Author: Short term we plan to aggregate violations. Long term is up in the air. There are some use cases where short-circuiting seems desirable (failed authz check should probably conceal any other validation errors..). I've made a note in future work.

  [`range`](https://github.com/google/cel-policy-templates-go/blob/master/test/testdata/map_ranges/template.yaml)
  feature that allows a CEL expression to be scoped to each entry of a map or
  item of an array. Multiple ranges can be combined to traverse a complex
  object.
- Kyverno [foreach
  declarations](https://kyverno.io/docs/writing-policies/validate/#foreach), use
  [JMESPath](https://jmespath.org/) to query for the elements that are then
  validated.

Note: We considered extending to a list of scopes, e.g.:

```yaml
  validations:
    - scopes: ["spec.containers[*]", "initContainers[*]", "spec.ephemeralContainers[*]"]
      expression: "scope.name.startsWith('xyz-')"
      message: "{1} does not start with 'xyz'"
      messageArgs: ["scope.name"]
```

But feedback was this is signficantly more difficult to understand.

### Message formatting alternatives

Alternative: offer a CEL expression

```yaml
- expression: "..."
  messageExpression: "string(object.int1) + ' is less than ' + string(object.int2)"
```

Cons:

- CEL requires explicit casts to string
- Plain string message support is a bit messy. Options:
  - Use CEL always: `"'this is a simple message'"`

>   - Offer both plain string (`message`) and CEL (`messageExpression`)

Reviewer 9: I actually like this option of two fields. And, I think the introduction of a `string.format()` function in CEL would address the first con of requiring explicit casts.

Author: If we can get `string.format()` in by beta I think I'm comfortable with this approach. We can add `messageExpression` to CRD validation as well.

apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  paramSource:
    group: rules.example.com
    kind: ReplicaLimit
    version: v1
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  denyReason: Invalid
  validations:
    - name: max-replicas
      expression: "object.spec.replicas <= params.maxReplicas"
      messageExpression: "'object.spec.replicas must be no greater than ' + string(params.maxReplicas)"
      # ...other rule related fields here...
```

The `spec.paramSource` field of the `ValidatingAdmissionPolicy` specifies the
kind of resources used to parameterize this policy. For this example, it is
configured by `ReplicaLimit` custom resources. Note in this example how the CEL
expression references to the parameters via the CEL `params` variable, e.g.
`params.maxReplicas`.

`spec.matchConstraints` specifies what resources this policy is designed to
validate. This also guides type-checking, see the "Informational type checking"
section for details.

> `denyReason` assigns a status code the result of this policy if it evaluates to

Reviewer 12: This suggests that if you need multiple reasons, you need multiple policies.  A common pattern in baked-in validation is

```
if field is nil {
   return "required" error
} else if field is formatted badly {
   return "invalid" error
}
```

Would it make sense to structure this like:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  validations:
    - name: field-x-specified
      denyReason: Required
      expression: "object.spec.x != null"
      messageExpression: "'object.spec.x must be provided"
    - name: field-x-valid
      denyReason: Invalid
      expression: "object.spec.x <= 42"
      messageExpression: "'object.spec.x must be greater than 42"
```

That would allow almost transliteration of baked-in validation (furthering the goal of everything being CRDs :)

Reviewer 12: Further refinement.

Built-in validation adds a lot of scaffolding for people to get mostly consistent (and parlty i18n-able) error messages for similar errors.  We do this with `field.Invalid()` and `field.Required()` and so on, which produce distinct structures.  What about something similar here, like:

```yaml
  validations:
    - name: field-x-specified
      expression: "object.spec.x != null"
      failure:
          required:
             fieldPath: "spec.x"
    - name: field-x-valid
      expression: "object.spec.x <= 42"
      failure:
          invalid:
             fieldPath: "spec.x"
             reason: "must be greater than 42"
```

We only need a small number of them (less than 10 in built-in code).  Heck, we could even enumerate common cases like

```yaml
      failure:
          invalid:
             fieldPath: "spec.x"
             greaterThan:
                 comparator:
                     literal: 42
```

Reviewer 12: I see this touched on later.  +1 on more structure and less free-form text, if our goal is to make good, consistent errors.  It takes constant vigilance with the builtin code already.

Author: Looks like we agree that `denyReason` should be per-validation.  There was one major KEP update not captured in this PR that fixed it, here's [where things landed](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3488-cel-admission-control/README.md#decisions-and-enforcement):

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  ...
  validations:
    - expression: "self.name.startsWith('xyz-')"
      name: name-prefix
      messageExpression: "self.name + ' must start with xyz-'"
      reason: Unauthorized
    - expression: "self.name.contains('bad')"
      name: bad-name
      message: "name contains 'bad' which is discouraged due to ..."
      code: 400
      reason: Invalid
    - expression: "self.name.contains('suspicious')"
      name: suspicious-name
      messageExpression: "self.name + ' contains suspicious'"
      code: 400
      reason: Invalid
```

Author: `fieldPath` is a really good idea. Big +1 for anything that establishes consistent structure.

We mostly punted to alpha2 on how expression failures translate to enforcement actions and how those are all represented. This will be something (the main thing really) that we circle back to this fall as we prep for 1.27 development.

I'll make note (via my next KEP update) of the idea to somehow incorporate the consistent structure of errors that `field.Invalid()` et. al. provide. I like the idea of taking a set of error categories that we've refined over a long time horizon and leveraging them here.

is being enforced, just in different ways, it is reasonable for both to share
the same policy configuration resources.

Alternative: `/matchRules` subresource

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: replicalimit.example.com
spec:
  group: example.com
  versions:
    ... # configuration schema(s) go here
  subresources:
    matchRules:
      matchRulesPath: .spec.match
```

Pros:

> - Follows pattern used by scale subresource to provide polymorphism across CRDs

Reviewer 3: IIRC subresources like /scale and /status are also intended to provide the ability to access/modify that subresource independently. I don't think those capabilities are a requirement here. I don't suppose there's any way to use that to streamline things?

Would shoehorning a subresource-like thing into a CRD just for the sake of injecting a schema and ignoring the rest of the features fragment the notion of subresources?

Are there other KEPs that could leverage "subresource-lite"?

Author: I agree that this a very different usage of subresource than the existing kubernetes subresources.

The closest thing to "subresource-lite" might be `Content-Type: PartialObjectMetadata` which is a way to read only the metdata of kubernetes resources.  Note that there is no subresource for metadata. `PartialObjectMetadata` is unusual in that it provides access to a part of resources that already has a well defined structure, so it is not a particularly good match for what we're trying to do either.

I'll list what you mentioned here in cons.

While webhook match rules give a good sense of what types of capabilities might
be needed, they serve a slightly different purpose.  Webhook match rules make it
possible to avoid webhook requests, which incur latency and impact availability,
for resources that don't need to be evaluated by the webhook. CEL expressions
have a comparatively small impact of latency and are in-process (and so do not
have the same impact to availability).

For CEL expressions, the primary benefits of match criteria are:

- Match criteria establishes bounds on what sort of admission requests the CEL
  expressions must consider. The CEL expressions can be written knowing that the
  match criteria has filtered out requests that is does not need to consider.
- Match criteria is available on policy configurations and allows the
  configuration author to further constrain what resources the particular
  configuration applies to.

Matching is performed in quite a few systems across Kubernetes:

| Match type                               | Usages in existing matchers  | Support?                       |
| ---------------------------------------- | ---------------------------- | ------------------------------ |

> | namespace selectors                      | WH/Audit/P&F                 | phase 1                        |

Reviewer 3: some G8r users have leveraged prefix/suffix matching for namespace/name

Author: For those use cases, would doing the prefix check in CEL be acceptable? I.e. use wider match criteria and then have the CEL expression allow resources where the namespace/name don't match the prefix/suffix?

Author: I've tracked this on the KEP with a UNRESOLVED and a link back to this thread.

| object selectors                         | WH                           | phase 1                        |

> | resource apiGroup + resource             | WH/Audit/P&F/RBAC            | phase 1                        |

Reviewer 3: thoughts on resource vs. kind?

Author: Advantages to both.

Both the replicaSet/scale and deployment/scale resources are the same Kind.  Resource matching can elect to just match on of these. Kind matching can match all resources of a Kind uniformly.

TBH, other than scale, there don't seem to be many examples of the same kinds being used across multiple resources. So resource and kind end up being interchangeable for large portions of the API surface.

Resource matching can match paths that have no kind, e.g. the /exec subresource.

Built-in Kubernetes matchers appear to all operate on resources.

Reviewer 3: Might be worth having matchers for both then? Would give policy frameworks the ability to use either.

Author: I'll propose having both in the KEP. This might get extra discussion later on, but I have a personal preference to matching on GVK for most use cases..

}
```

`MatchPolicy` will work the same as for admission webhooks. It will default to
`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
definitions for different versions of CRD" for an explanation of why we need
`MatchPolicy`.

xref:
https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules

Alternative considered: Define match criteria with CEL expressions

Pros:

- CEL expressions can be used to declare far more sophisticated match criteria.

Cons:

- CEL expressions can evalaute to an error while API types only evaluate to
  "matches" / "does not match" result.

> - We intend to allow policy configuration resources declare match criteria,

Reviewer 3: may also make language-agnosticism more difficult to achieve in the future (no ability to add a "language" siblling field as contemplated in Brendan Burns's comment.

Author: Very good point. Added to cons.

  making it difficult to pre-compile or type-check CEL expressions.

##### Exclude matching

Benefits:

- Exclude rules make it easier for policy definition authors to do things like
  exclude kube-system or other special purpose namespaces from policies that apply
  to all "normal" namespaces.
- Default rules make it easier to apply principal of least privilege to
  policies. I.e. ensure that newly created resources default to a least
  privilege state until configured otherwise.

> Exclude rules can be modeled after the approach exstablished in Audit where

Reviewer 3: I think there are at least two approaches to exclude: interactive and non-interactive. This interactive one is more problematic, and I don't see the non-interactive one considered.

This approach appears to describe a resource that can preempt other resources based off priority. This complicates the ability to reason about policy, since adding a new policy may not actually be effective due to priority, and also makes the eventual consistency story harder to reason about: in a world where policies don't interact then a non-ingested policy means your cluster is strictly less locked-down than specified, that same assumption can't be made if inter-policy interactions are added.

Another approach is to add exclude rules that apply only to the policy in-question e.g. `match.excludedNamespaces`, but don't affect other policies. This is essentially a denormalized version of what's currently suggested.

It may make sense to have global exclusion objects, per https://github.com/kubernetes/kubernetes/issues/92157 but that would be to remove control of certain resources from the user space entirely, so cluster admins can ensure their "superroot" operations can't be interfered with.

Author: I'm not familiar with the notion of interactive and non-interactive.

Unless I'm misunderstanding how Audit excludes work, the `level=None` has a priority (based on order in the match list), but only impacts the list of match rules contained in that Audit resource, not across resources.  I agree that any exclude rules that applies across priority resources is very problematic.

My mental model for what I want matching criteria to do is basically a query filter (e.g. SQL WHERE clause).  `(<some criteria>) AND (<some criteria>) AND NOT (<some exclude criteria>)`.

To use Audit match rules I somehow have to express this in a list of match criteria that short-circuits on the first match. For cases with very few exclude rules this seems manageable?  I'm open to ideas here. I'll admit I'm biasing toward using approaches already in use in the Kubernetes API, but I think if we find them deficient, we should do something better.

Author: https://github.com/kubernetes/enhancements/pull/3492#discussion_r970211846

level=None on a match rule can be used to exlude any rules beneath it. E.g.
`effect: Exclude`.

If we want this, it is essential that we introduce it early, the exclude inverts
the semantics of a match, adding it in a later release is _very_ incompatible.

> #### Singleton policies

Reviewer 3: Adding this complicates reporting/debugging/analyzing the policy space somewhat. To the extent that this is purely a UX benefit (I don't think it allows expression of anything otherwise unexpressable), the downsides should also be considered.

Author: This seems manageable. If we report most violations using a {policy definition, policy configuration} identifier pair (presumably names since they are both cluster scope in phase 1), then for singletons we could just use {policy definition, policy definition} as the identifier pair. It keeps things consistent and makes tracing from violation info back to the resource that produced it straight-forward.

Note that this does allow `ValidatingAdmissionPolicy` to match
`ValidatingWebhookConfiguration`.

Note: In the future we may further loosen this up and allow admission
configuration to intercept/guard writes to admission configuration while
preventing deadlock - Add feature to configure a set of webhooks to intercept
other webhooks https://github.com/kubernetes/kubernetes/issues/101794.

Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
higher level. This could be added later.

> #### Reporting violations

Reviewer 3: How does violation reporting interact with policy frameworks? For example, Gatekeeper also reports the # of webhook violations as a prometheus metric. If the goal is to have higher-level frameworks provide a single pane of glass (e.g. they are also polyfilling the stuff that embedded policy wont/can't do), then some kind of two-way communication is needed.

If we are okay with fragmented reporting, then we probably should restate how this KEP is positioning itself WRT extensibility.

Author: This should be titled "Reporting violations back in response to client". We will need metrics.

Author: Clarified and added a separate metrics section.

Goals:

- Feature parity with `AdmissionReview`
- Ability to format message strings
- support for reasons/codes

Example:

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  ...
  validations:
    - expression: "self.name.startsWith('xyz-')"

>       level: Deny # the default

Reviewer 3: Different enforcement points may want different enforcement actions and may have different possible actions they can take. If the goal is to have this syntax shift left/right, we want to tread carefully here.

Another option would be to let policy frameworks handle the higher-level consideration and assume this is purely for admission-time enforcement.

Author: I _really_ want to put enforcement action under the control of the policy configurations.

Things that come to mind:
- Should each validation expression in a policy be able to have it's own enforcement action? (or should we just agree that all the validation expressions of policy share enforcement action?)
- What happens two policy definitions share a configuration, seems like the only option is to have the configuration's enforcement action effect both policies. How does one roll out a policy in this model?

Author: I've tried moving enforcement to the configurations by requiring names for all validation expressions and then requiring that policy configurations set the level of each validation expression (this is also useful for audit which needs a key for each violation).

Author: Moved enforcement control to policy configuration.

      messageExpression: "self.name + ' must start with \"xyz-\"'"
      reason: Invalid
    - expression: "self.name.contains('bad')"
      level: Warn
      message: "name contains 'bad' which is discouraged due to ..."
    - expression: "self.name.contains('suspicious')"

>       level: Audit

Reviewer 3: Setting the enforcement level in the policy definition potentially makes it impossible for an admin to customize the level (unless the definition author parameterizes it). This makes it more challenging for an admin to have different severities in different scopes enforced via match criteria.

Use-case example: gradually rolling out a new policy, switching it from `Warn` to `Deny` namespace-by-namespace as pre-existing violations are resolved.

Author: I wonder if the both policy and the configurations should have some control of enforcement level? See my comments in above thread.

Author: Moved enforcement control entirely into policy configuration.

      messageExpression: "self.name + ' contains \"suspicious\"'"
      auditKey: suspicious-container-name
```

Equivalent `AdmissionReview`:

```yaml
request:
  ...
response:
  allowed: false
  status:
    #status: Failure
    message: "nothing-suspicious-or-bad-here must start with \"xyz-\""
    reason: Invalid
    code: 422 # Code for "Invalid"
  auditAnnotations:
    suspicious-container-name: "nothing-suspicious-or-bad-here contains \"suspicious\""
  warnings:
    - "name contains \"bad\" which is discouraged due to ..."
```

The resulting message would be something like:

```
configuration validate-config1.example.com of admission policy validate-xyz.example.com denied the request: nothing-suspicious-or-bad-here must start with "xyz-"
```

Notes:

> - using `level` makes it easy to toggle between `Warn` and `Deny`, it also

Reviewer 3: note: the "level" syntax is presuming a specific format for CEL, I don't think we've really sussed out what we want there.

Author: This part still needs work. Parity with webhook admission responses requires that we support accept/deny, multiple warnings and multiple audit annotations (each with an audit key). How we express this in relation to CEL expressions is where this gets interesting.

Author: I've dropped "level", there are different enforcement options (channels really), more than one can be active at a time.

  makes `Audit` more consistent with the other outputs (although it does require
  a key field).
- `message` is for plain string messages that do not need any formatting
  (`messageExpression` and `message` would be exclusive)
- if there are not `Deny` validations, `response.allowed` will automatically be `true`

TODOs:

- What if multiple validations fail?  What if they have different reasons (Invalid + Forbidden)?

>   - Current admission chain short circuits on validating admission rejections? So just return 1st failed?

Reviewer 3: This could be annoying for users who then have to discover violations one-by-one

Author: I agree. We should be able to support this. I've made a note.

  "HOLD" value.
- Validate that only a controller responsible for a finalizer can remove it from
  the finalizers field.

This could be supported by matching criteria, or via CEL expression access, or both.
 
 Concerns: "Is joe authorized to do this"? That only works for the objects joe
 creates, but not objects that get created on joe's behalf by a controller.
 Ditto for updates. I heard someone cite PSP as an example for why it's needed,
 but IMO that was an anti-pattern of PSP, and one that we explicitly decided to
 omit from PSA

> #### Default matching

Reviewer 3: What is the use case behind default matching?

What is the behavior if a user defines match criteria:

* on the policy definition (not relevant for some of these alternatives)
* on the policy config object

Is the default matching overridden or evaluated in addition to the user-specified match criteria

Author: Primary use case is defining a least privilege policy, e.g. "Restrict maxReplicas to 5 unless there is a policy configuration that sets a different limit" or "Do not allow any of the following label keys to be used unless there is a policy configuration allowing them for the resource being admitted".

For these cases to work, the default matching would on apply if no policy configuration match.

There are cases where the desired effect doesn't require feature support and can just have a normal policy configuration that covers the "default case", e.g. If you have a global policy configuration for some upper limit to maxReplicas (say 1000), you could also have lower limits for scoped policy configurations without needing this "default configuration" feature.  Or, you can use namespace label selectors (if the namespace's environment label not in ['test', 'staging', 'prod']) to define what the default configuration should match.

The idea here is to generalize support for least privilege policy scenarios. Useful? Could be added later safely?

Author: I found a simpler way to incorporate this. See "Default Validations" in latest KEP version.

the Kubernetes Go structs used to representative native API types, both for type
checking and for runtime data access.

### Writing to Status

This enhancement proposes using status to of `ValidatingAdmissionPolicy` to
communicate type-checking errors and any other misconfigurations such as CRD not
found errors.

As mentioned in
https://github.com/kubernetes/enhancements/pull/3492#discussion_r964841045,
status on API server configuration objects has been tricky to design in the
past, because of the following:

- multiple active kube-apiservers (sometimes at identical versions, sometimes
  skewed by one version during upgrade)
- multiple active non-kube-apiserver servers (aggregated servers)

As a concrete example, The CRD `NonStructural` status field takes advantage of
the metadata generation field
(https://github.com/kubernetes/kubernetes/commit/2cfc3c69dc7c17b2711af0168f39ed7f515675c2).
This avoids repeated updates for the same generation and potential fights of API

> server in HA environments even without leader elected controllers.

Reviewer 3: If the controllers are not singletons, can there ever truly be one status?

Imagine a multi API server cluster is mid-upgrade and a resource gets created that leverages a field acknowledged in the newer version, but not the older version.

If 1/3 of the API servers are on the newer version, then that server will report "everything's looking good", the other two servers will report "error".

In this situation, enforcement will be random depending on which API server is hit, which is something a user may be interested in.

Reviewer 3: The second half of this kubecon talk addresses status in a multi-controller/webhook environment:

https://sched.co/ekBY

Author: Not truely. 

And even when the controller is leader elected, it is going to write to status based on the Kubernetes version is communicates with so the same problem exists.

Even when all the apiservers are on the same version, one might be seeing a newer revision of a CRD than the others, resulting in a similar case.

Also, when a webhook is registered, the is no bound on how long it will take for all API servers to realize it exists and include it in the admission chain. And the apiserver currently make no attempts to show in the status of CRDs which apiservers they are registered (or which resourceVersion that apiserver is using). Same for webhooks.

We will use a similar approach. The complication here is that for this case we
must consider up to 3 resources:

- `ValidatingAdmissionPolicy` resource
- Configuration CRD

> - If policy matches a CRD, the matched CRD

Reviewer 3: "matches a CRD" == "matches a configuration resource"?

Reviewer 3: Or is this the CRD that defines the kind-under-test, to be used for type checking?

Author: Yes, I'll rephrase.
